<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Coralysis sensitive identification of imbalanced cell types and states in single-cell data via multi-level integration • Coralysis</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Coralysis sensitive identification of imbalanced cell types and states in single-cell data via multi-level integration">
<meta name="description" content="Coralysis is an R package featuring a multi-level integration algorithm for sensitive integration, reference-mapping, and cell state identification in single-cell data.">
<meta property="og:description" content="Coralysis is an R package featuring a multi-level integration algorithm for sensitive integration, reference-mapping, and cell state identification in single-cell data.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">Coralysis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/01_QuickStart.html">Integration &amp; reference-mapping</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/elolab/Coralysis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="coralysis">Coralysis<a class="anchor" aria-label="anchor" href="#coralysis"></a>
</h1></div>
<p><br></p>
<div class="section level4">
<h4 id="coralysis-is-an-r-package-featuring-a-multi-level-integration-algorithm-for-sensitive-integration-reference-mapping-and-cell-state-identification-in-single-cell-data-described-in-the-paper-coralysis-enables-sensitive-identification-of-imbalanced-cell-types-and-states-in-single-cell-data-via-multi-level-integration">Coralysis is an R package featuring a multi-level integration algorithm for sensitive integration, reference-mapping, and cell state identification in single-cell data, described in the paper: <em><a href="https://doi.org/10.1101/2025.02.07.637023" class="external-link">Coralysis enables sensitive identification of imbalanced cell types and states in single-cell data via multi-level integration</a></em>.<a class="anchor" aria-label="anchor" href="#coralysis-is-an-r-package-featuring-a-multi-level-integration-algorithm-for-sensitive-integration-reference-mapping-and-cell-state-identification-in-single-cell-data-described-in-the-paper-coralysis-enables-sensitive-identification-of-imbalanced-cell-types-and-states-in-single-cell-data-via-multi-level-integration"></a>
</h4>
<p><br></p>
<p><br></p>
<p><img src="articles/Coralysis_flowchart.png"></p>
<p><em>Coralysis integration flowchart. (A) An input of heterogeneous scRNA-seq datasets are overclustered batch wise into a training set modelled through the Iterative Clustering Projection (ICP) algorithm in order to predict the cell cluster probabilities and obtain an integrated embedding. Adaptations to the original ICP algorithm (Smolander et al., 2021): (B) batch wise cluster assignment at start, dependent on the cell distribution across Principal Component 1 (median as cutoff); (C) training cells selected from batch k nearest neighbours of the cell with the highest probability for every batch per cluster; and, (D) upon ICP clustering convergence, each cluster is further divided into two for the next clustering round, dependent on the batch wise cluster probability distribution (median as cutoff). (E) Multi-level integration is achieved through multiple divisive clustering rounds, blending the batch effect and highlighting the biological signal incrementally. Shapes represent cell types and colours batches.</em></p>
<p><br></p>
<hr>
<p><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><br></p>
<p>The latest version of <code>Coralysis</code> can be downloaded from GitHub using the devtools R package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"elolab/Coralysis"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<hr>
<p><br></p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick-start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p><br></p>
<p>To run <code>Coralysis</code>, follow the instructions below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/elolab/Coralysis" class="external-link">"Coralysis"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Prepare data: checks 'logcounts' format &amp; removes non-expressed genes</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/PrepareData.html">PrepareData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Multi-level integration</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RunParallelDivisiveICP.html">RunParallelDivisiveICP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, batch.label <span class="op">=</span> <span class="st">"batch"</span>, threads <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Compute integrated PCA from joint probability</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">125</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RunPCA.html">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run UMAP</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1204</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, umap.method <span class="op">=</span> <span class="st">"uwot"</span>, n_neighbors <span class="op">=</span> <span class="fl">20</span>, min_dist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize integrated data </span></span>
<span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"batch"</span>, <span class="st">"cell_type"</span><span class="op">)</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">vars</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, color.by <span class="op">=</span> <span class="va">x</span>, point.size <span class="op">=</span> <span class="fl">0.5</span>, point.stroke <span class="op">=</span> <span class="fl">0.5</span>, legend.nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> </span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">2</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span> <span class="co"># join plots together</span></span></code></pre></div>
<p><br></p>
<hr>
<p><br></p>
</div>
<div class="section level2">
<h2 id="contact-information">Contact information<a class="anchor" aria-label="anchor" href="#contact-information"></a>
</h2>
<p><br></p>
<p>If you have questions related to <code>Coralysis</code>, please contact us <a href="https://github.com/elolab/Coralysis/issues" class="external-link">here</a>.</p>
<p><br></p>
<hr>
<p><br></p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p><br></p>
<p>Sousa A, Smolander J, Junttila S, Elo L (2025). <strong>Coralysis enables sensitive identification of imbalanced cell types and states in single-cell data via multi-level integration.</strong> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2025.02.07.637023" class="external-link">doi:10.1101/2025.02.07.637023</a></p>
<p><br></p>
<hr>
<p><br></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p><br></p>
<ol style="list-style-type: decimal">
<li>Johannes Smolander, Sini Junttila, Mikko S Venäläinen, Laura L Elo. “ILoReg: a tool for high-resolution cell population identification from single-cell RNA-seq data”. Bioinformatics, Volume 37, Issue 8, 15 April 2021, Pages 1107–1114, <a href="https://doi.org/10.1093/bioinformatics/btaa919" class="external-link">https://doi.org/10.1093/bioinformatics/btaa919</a>.</li>
</ol>
<p><br></p>
<p><br></p>
<p><br></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/elolab/Coralysis/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/elolab/Coralysis/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Coralysis</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>António Sousa <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0003-4779-6459" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Johannes Smolander <br><small class="roles"> Contributor, author </small> <a href="https://orcid.org/0000-0003-3872-9668" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Sini Junttila <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3754-5584" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Laura L Elo <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-5648-4532" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by António Sousa, Johannes Smolander, Sini Junttila, Laura L Elo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
