---
title: "ILoReg2: sensitive cell population identification through multi-level integration of single-cell RNA-seq data"
output: html_document
date: "2024-08-27"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Integration

```{r packages}
# Packages
library("SingleCellExperiment")
library("ILoReg2")
library("scater")
library("ggplot2")
#library("scran") 
#library("cowplot")
```

```{r data}
# Import data distributed with ILoReg2
data("pbmc_10Xassays") # it imports object as sce
pbmc_10Xassays
```

Check how the data looks before integrating. 

```{r unintegrated, fig.width=7, fig.height=2.5}
# Compute PCA & TSNE
pbmc_10Xassays <- scran::fixedPCA(pbmc_10Xassays, rank = 30, subset.row = NULL)
set.seed(123)
pbmc_10Xassays <- RunTSNE(pbmc_10Xassays)

# Plot TSNE highlighting the batch & cell type
unint.batch.plot <- plotTSNE(pbmc_10Xassays, colour_by = "batch", point_size = 0.01) + 
    guides(color = guide_legend(override.aes = list(size=2.5))) # to increase the legend dot size
unint.cell.plot <- plotTSNE(pbmc_10Xassays, colour_by = "cell_type", point_size = 0.01) + 
    guides(color = guide_legend(override.aes = list(size=2.5)))
cowplot::plot_grid(unint.batch.plot, unint.cell.plot, rel_widths = c(0.45, 0.55))
```

Run integration with `ILoReg2`. Provide the batch label column name in `colData(pbmc_10Xassays)` to the parameter `batch.label`.  

```{r integration, fig.width=7, fig.height=2.5}
# Perform integration with ILoReg2
# start <- Sys.time()
set.seed(1024)
pbmc_10Xassays <- RunParallelDivisiveICP(object = pbmc_10Xassays, batch.label = "batch", threads = 4)
# stop <- Sys.time()
# print(stop - start) # Time difference of 11.41677 mins

# Compute PCA with joint cluster probabilities & TSNE
pbmc_10Xassays <- RunPCA(pbmc_10Xassays, scale = T)
set.seed(123)
pbmc_10Xassays <- RunTSNE(pbmc_10Xassays)

# Plot TSNE highlighting the batch & cell type
int.batch.plot <- plotTSNE(pbmc_10Xassays, colour_by = "batch", point_size = 0.01) + 
    guides(color = guide_legend(override.aes = list(size=2.5)))
int.cell.plot <- plotTSNE(pbmc_10Xassays, colour_by = "cell_type", point_size = 0.01) + 
    guides(color = guide_legend(override.aes = list(size=2.5)))
cowplot::plot_grid(int.batch.plot, int.cell.plot, rel_widths = c(0.45, 0.55))
```

## High-resolution clustering

```{r clustering, fig.width=9, fig.height=2.5}
# Clustering 
pbmc_10Xassays <- HierarchicalClustering(object = pbmc_10Xassays)

# Select no. of clusters
pbmc_10Xassays <- SelectKClusters(pbmc_10Xassays, K = 16)

# Plot clustering
clt.plot <- ClusteringScatterPlot(pbmc_10Xassays, dim.reduction.type = "tsne", return.plot = TRUE, 
                                  point.size = 0.01, show.legend = FALSE, title = "ILoReg2 clusters")
cowplot::plot_grid(int.batch.plot, int.cell.plot, clt.plot, ncol = 3, 
                   rel_widths = c(0.325, 0.375, 0.275), align = "h")
```

## Cluster markers

```{r cluster markers, fig.width=5, fig.height=5}
# Cluster markers 
cluster.markers <- FindAllGeneMarkers(pbmc_10Xassays)

# Select the top 3 positive markers per cluster 
top3.markers <- SelectTopGenes(cluster.markers, top.N = 3,  criterion.type = "log2FC", inverse = FALSE)

# Heatmap of the top 3 positive markers per cluster
GeneHeatmap(pbmc_10Xassays, gene.markers = top3.markers)
```

## DGE

```{r dge}
# DGE analysis: cluster 2 vs 6
dge.clt2vs6 <- FindGeneMarkers(pbmc_10Xassays, clusters.1 = "2", clusters.2 = "6")
head(dge.clt2vs6)
```


## Reference-mapping

Perform reference-mapping with `ILoReg2`, 

```{r reference-mapping, fig.width=16, fig.height=3}
## Reference-mapping

# Split the SCE object by 'batch'
ref <- pbmc_10Xassays[,pbmc_10Xassays$batch=="V2"] # let V2 assay batch be the reference data set
query <- pbmc_10Xassays[,pbmc_10Xassays$batch=="V1"] # let V1 be the query (unknown annotations)

# 1) Train the reference 
# start <- Sys.time()
set.seed(123)
ref <- RunParallelDivisiveICP(object = ref, divisive.method = "cluster", 
                              threads = 4) # runs without 'batch.label' as it represents 1 sample only
# stop <- Sys.time()
# print(stop - start) # Time difference of 8.052504 mins

# 2) Compute reference PCA & TSNE
ref <- RunPCA(ref, scale = T, return.model = TRUE, method = "stats")
set.seed(123)
ref <- RunUMAP(ref, return.model = TRUE)

# 3) Project & predict query cell labels 
# start <- Sys.time()
set.seed(1024)
map <- ReferenceMapping(ref = ref, query = query, ref.label = "cell_type",
                        scale.query = FALSE, project.umap = TRUE)
# stop <- Sys.time()
# print(stop - start) # Time difference of 2.380615 mins

# Calculate accuracy
preds_x_truth <- table(map$ilo_labels, map$cell_type)
preds_x_truth
sum(diag(preds_x_truth)) / sum(preds_x_truth) # 0.848847 : 85%

# Plot query and reference UMAP side-by-side with ground-truth & predicted cell labels
ref.celltype.plot <- plotUMAP(ref, colour_by = "cell_type", point_size = 0.01) + 
    guides(color = guide_legend(override.aes = list(size=2.5))) + ggtitle("reference (ground-truth)")
query.ground_truth.plot <- plotUMAP(map, colour_by = "cell_type", point_size = 0.01) + 
    guides(color = guide_legend(override.aes = list(size=2.5))) + ggtitle("query (ground-truth)")
query.predicted.plot <- plotUMAP(map, colour_by = "ilo_labels", point_size = 0.01) + 
    guides(color = guide_legend(override.aes = list(size=2.5))) + ggtitle("query (predicted)")
query.confidence.plot <- plotUMAP(map, colour_by = "ilo_labels_prob", point_size = 0.01) + 
    guides(color = guide_legend(override.aes = list(size=2.5))) + ggtitle("query (confidence)")
cowplot::plot_grid(ref.celltype.plot, query.ground_truth.plot, query.predicted.plot, query.confidence.plot, ncol=4)
```

```{r rsession}
# R session
sessionInfo()
```
