<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Integration • Coralysis</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integration">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Coralysis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/Coralysis.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/01_Integration.html">Integration</a></li>
    <li><a class="dropdown-item" href="../articles/02_RefMap.html">Reference-mapping</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/elolab/Coralysis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Integration</h1>
            
            <h4 data-toc-skip class="date">Compiled: 13/02/2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/elolab/Coralysis/blob/main/vignettes/01_Integration.Rmd" class="external-link"><code>vignettes/01_Integration.Rmd</code></a></small>
      <div class="d-none name"><code>01_Integration.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p><br></p>
<p>To illustrate the multi-level integration algorithm, we will use in
this vignette two 10X PBMCs (Peripheral Blood Mononuclear Cells) 3’
assays: <strong>V1</strong> and <strong>V2</strong>. The datasets have
been downloaded from 10X website. The PBMC dataset <strong>V1</strong>
corresponds to sample <em>pbmc6k</em> and <strong>V2</strong> to
<code>pbmc8k</code>:</p>
<ul>
<li><p><strong>V1</strong>: <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc6k" class="external-link">pbmc6k</a></p></li>
<li><p><strong>V2</strong>: <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k" class="external-link">pbmc8k</a></p></li>
</ul>
<p><br></p>
<p>Cells were annotated using the annotations provided by <a href="https://www.nature.com/articles/s41592-019-0619-0" class="external-link">Korsunsky et
al., 2019</a> (Source Data Figure 4 file). The overall data was
downsampled to 2K cells (1K per assay) and 2K highly variable genes
selected with <code>scran</code> R package. To facilitate the
reproduction of this vignette, the data is distributed through
<em>Zenodo</em> as a <code>SingleCellExperiment</code> object, the
object (class) required by most functions in <code>Coralysis</code> (see
<a href="https://bioconductor.org/books/3.13/OSCA.intro/the-singlecellexperiment-class.html" class="external-link">Chapter
4 The SingleCellExperiment class - OSCA manual</a>). The
<code>SCE</code> object provided comprises <code>counts</code> (raw
count data), <code>logcounts</code> (log-normalized data) and cell
<code>colData</code> (which includes batch and cell labels, designated
as <code>batch</code> and <code>cell_type</code>, respectively).</p>
<p>Run the code below to import the <code>R</code> packages and data
required to reproduce this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/elolab/Coralysis" class="external-link">"Coralysis"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Import data from Zenodo</span></span>
<span><span class="va">data.url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/14845751/files/pbmc_10Xassays.rds?download=1"</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="va">data.url</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="dimred-pre-integration">DimRed: pre-integration<a class="anchor" aria-label="anchor" href="#dimred-pre-integration"></a>
</h2>
<p><br></p>
<p>The batch effect between assays can be inspected below by projecting
the data onto <em>t</em>-distributed Stochastic Neighbor Embedding
(t-SNE). This can be achieved by running sequentially the
<code>Coralysis</code> functions <code>RunPCA</code> and
<code>RunTSNE</code>. Provide a seed before running each one of these
functions to ensure reproducibility. The function <code>RunPCA</code>
runs by default the PCA method implemented the R package
<code>irlba</code> (<code>pca.method="irlba"</code>), which requires a
seed to ensure the same PCA result. In addition, the
<code>assay.name</code> argument needs to be provided, otherwise uses by
default the probabilities which are obtained only after integration
(after running <code>RunParallelDivisiveICP</code>). The assay
<code>logcounts</code>, corresponding to the log-normalized data, and
number of principal components to use <code>p</code> were provided. In
this case, the data has been previously normalized, but it could have
been normalized using methods available in <em>Bioconductor</em> (see <a href="https://bioconductor.org/books/3.12/OSCA/normalization.html" class="external-link">Chapter
7 Normalization - OSCA manual</a>). Any categorical variable available
in <code>colData(pbmc_10Xassays)</code>, such as <code>batch</code> or
<code>cell_type</code>, can be visualized in a low dimensional embedding
stored in <code>reducedDimNames(pbmc_10Xassays)</code> with the
<code>Coralysis</code> function <code>PlotDimRed</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute PCA &amp; TSNE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunPCA.html">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, </span>
<span>                         assay.name <span class="op">=</span> <span class="st">"logcounts"</span>, </span>
<span>                         p <span class="op">=</span> <span class="fl">30</span>, dimred.name <span class="op">=</span> <span class="st">"unintPCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunTSNE.html">RunTSNE</a></span><span class="op">(</span><span class="va">pbmc_10Xassays</span>, </span>
<span>                          dimred.type <span class="op">=</span> <span class="st">"unintPCA"</span>, </span>
<span>                          dimred.name <span class="op">=</span> <span class="st">"unintTSNE"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot TSNE highlighting the batch &amp; cell type</span></span>
<span><span class="va">unint.batch.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, </span>
<span>                               color.by <span class="op">=</span> <span class="st">"batch"</span>, </span>
<span>                               dimred <span class="op">=</span> <span class="st">"unintTSNE"</span>,</span>
<span>                               point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                               legend.nrow <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                               seed.color <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span></span>
<span><span class="va">unint.cell.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, </span>
<span>                              color.by <span class="op">=</span> <span class="st">"cell_type"</span>, </span>
<span>                              dimred <span class="op">=</span> <span class="st">"unintTSNE"</span>, </span>
<span>                              point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                              legend.nrow <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                              seed.color <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">unint.batch.plot</span>, <span class="va">unint.cell.plot</span>, ncol <span class="op">=</span> <span class="fl">2</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span></span></code></pre></div>
<p><img src="01_Integration_files/figure-html/dimred:%20pre-integration-1.png" width="576" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="multi-level-integration">Multi-level integration<a class="anchor" aria-label="anchor" href="#multi-level-integration"></a>
</h2>
<p><br></p>
<p>Integrate assays with the multi-level integration algorithm
implemented in <code>Coralysis</code> by running the function
<code>RunParallelDivisiveICP</code>. The only arguments required by this
function are <code>object</code> and <code>batch.label</code>. The
<code>object</code> requires a <code>SingleCellExperiment</code> object
with the assay <code>logcounts</code>. The matrix in
<code>logcounts</code> should be sparse, i.e.,
<code>is(logcounts(pbmc_10Xassays), "dgCMatrix")</code> is
<code>TRUE</code>, and it should not contain non-expressing genes. This
is ensured by running <code>PrepareData</code> before. The
<code>batch.label</code> argument requires a label column name in
<code>colData(pbmc_10Xassays)</code> corresponding to the batch label
that should be used for integration. In the absence of a batch, the same
function, <code>RunParallelDivisiveICP</code>, can be run without
providing <code>batch.label</code> (i.e.,
<code>batch.label = NULL</code>), in which case the data will be modeled
through the algorithm to identify fine-grained populations that do not
required batch correction. An higher number of <code>threads</code> can
be provided to speed up computing time depending on the number of cores
available. For this example, the algorithm was run 10 times
(<code>L = 10</code>), but generally, this number should be higher (with
the default being <code>L = 50</code>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prepare data for integration: </span></span>
<span><span class="co">#remove non-expressing genes &amp; logcounts is from `dgCMatrix` class</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PrepareData.html">PrepareData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Data in `logcounts` slot already of `dgCMatrix` class...</span></span></code></pre>
<pre><code><span><span class="co">## 2000/2000 features remain after filtering features with only zero values.</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform integration with Coralysis</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1024</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunParallelDivisiveICP.html">RunParallelDivisiveICP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, </span>
<span>                                         batch.label <span class="op">=</span> <span class="st">"batch"</span>, </span>
<span>                                         L <span class="op">=</span> <span class="fl">10</span>, threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Building training set...</span></span></code></pre>
<pre><code><span><span class="co">## Training set successfully built.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Computing cluster seed.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Initializing divisive ICP clustering...</span></span></code></pre>
<pre><code><span><span class="co">##   |                                                                              |                                                                      |   0%  |                                                                              |========                                                              |  11%  |                                                                              |================                                                      |  22%  |                                                                              |=======================                                               |  33%  |                                                                              |===============================                                       |  44%  |                                                                              |=======================================                               |  56%  |                                                                              |===============================================                       |  67%  |                                                                              |======================================================                |  78%  |                                                                              |==============================================================        |  89%  |                                                                              |======================================================================| 100%</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Divisive ICP clustering completed successfully.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Predicting cell cluster probabilities using ICP models...</span></span></code></pre>
<pre><code><span><span class="co">## Prediction of cell cluster probabilities completed successfully.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multi-level integration completed successfully.</span></span></code></pre>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="dimred-post-integration">DimRed: post-integration<a class="anchor" aria-label="anchor" href="#dimred-post-integration"></a>
</h2>
<p><br></p>
<p>The integration result can be visually inspected by running
sequentially the functions <code>RunPCA</code> and <code>RunTSNE</code>.
The <code>assay.name</code> provided to <code>RunPCA</code> must be
<code>joint.probability</code> (the default), the primary output of
integration with <code>Coralysis</code>. The probability matrices from
<code>Coralysis</code> (i.e., <code>joint.probability</code>) can be
used to obtain an integrated embedding by running
<code>RunPCA(..., assay.name = "joint.probability")</code>. This
integrated PCA can, in turn, be used downstream for clustering or
non-linear dimensional reduction techniques, such as
<code>RunTSNE</code>. Below, the integrated PCA was named
<code>intPCA</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute PCA with joint cluster probabilities &amp; TSNE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">pbmc_10Xassays</span>, </span>
<span>                         assay.name <span class="op">=</span> <span class="st">"joint.probability"</span>, </span>
<span>                         dimred.name <span class="op">=</span> <span class="st">"intPCA"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Divisive ICP: selecting ICP tables multiple of 4</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunTSNE.html">RunTSNE</a></span><span class="op">(</span><span class="va">pbmc_10Xassays</span>, </span>
<span>                          dimred.type <span class="op">=</span> <span class="st">"intPCA"</span>, </span>
<span>                          dimred.name <span class="op">=</span> <span class="st">"intTSNE"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot TSNE highlighting the batch &amp; cell type</span></span>
<span><span class="va">int.batch.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, </span>
<span>                             color.by <span class="op">=</span> <span class="st">"batch"</span>, </span>
<span>                             dimred <span class="op">=</span> <span class="st">"intTSNE"</span>, </span>
<span>                             point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                             legend.nrow <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                             seed.color <span class="op">=</span> <span class="fl">1024</span><span class="op">)</span></span>
<span><span class="va">int.cell.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, </span>
<span>                            color.by <span class="op">=</span> <span class="st">"cell_type"</span>, </span>
<span>                            dimred <span class="op">=</span> <span class="st">"intTSNE"</span>, </span>
<span>                            point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                            legend.nrow <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                            seed.color <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">int.batch.plot</span>, <span class="va">int.cell.plot</span>, </span>
<span>                   ncol <span class="op">=</span> <span class="fl">2</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span></span></code></pre></div>
<p><img src="01_Integration_files/figure-html/dimred:%20post-integration-1.png" width="576" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<p><br></p>
<p>Run graph-based clustering with the <code>scran</code> function
<code>clusterCells</code> (see <a href="https://bioconductor.org/books/3.14/OSCA.basic/clustering.html" class="external-link">Chapter
5 Clustering - OSCA manual</a>).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Graph-based clustering on the integrated PCA w/ 'scran' package</span></span>
<span><span class="va">blusparams</span> <span class="op">&lt;-</span> <span class="fu">bluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">15</span>, cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">pbmc_10Xassays</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">pbmc_10Xassays</span>, </span>
<span>                                              use.dimred <span class="op">=</span> <span class="st">"intPCA"</span>, </span>
<span>                                              BLUSPARAM <span class="op">=</span> <span class="va">blusparams</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot clustering</span></span>
<span><span class="va">clt.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, </span>
<span>                       color.by <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>                       dimred <span class="op">=</span> <span class="st">"intTSNE"</span>, </span>
<span>                       point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                       legend.nrow <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                       seed.color <span class="op">=</span> <span class="fl">65</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">int.batch.plot</span>, <span class="va">int.cell.plot</span>, </span>
<span>                   <span class="va">clt.plot</span>, ncol <span class="op">=</span> <span class="fl">3</span>, align <span class="op">=</span> <span class="st">"h"</span><span class="op">)</span></span></code></pre></div>
<p><img src="01_Integration_files/figure-html/clustering-1.png" width="1056" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="cluster-markers">Cluster markers<a class="anchor" aria-label="anchor" href="#cluster-markers"></a>
</h2>
<p><br></p>
<p>Identify the cluster markers by running the <code>Coralysis</code>
function <code>FindAllClusterMarkers</code>. Provide the
<code>clustering.label</code>, in this case, the label used above, i.e.,
<code>cluster</code>. The top three positive markers per cluster were
retrieved and plotted below using the <code>Coralysis</code> function
<code>HeatmapFeatures</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster markers </span></span>
<span><span class="va">cluster.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindAllClusterMarkers.html">FindAllClusterMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, clustering.label <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 1</span></span>
<span><span class="co">## 1128 features left after min.pct filtering</span></span>
<span><span class="co">## 1128 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 215 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 2</span></span>
<span><span class="co">## 1203 features left after min.pct filtering</span></span>
<span><span class="co">## 1203 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 287 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 3</span></span>
<span><span class="co">## 1167 features left after min.pct filtering</span></span>
<span><span class="co">## 1167 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 427 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 4</span></span>
<span><span class="co">## 1171 features left after min.pct filtering</span></span>
<span><span class="co">## 1171 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 443 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 5</span></span>
<span><span class="co">## 1194 features left after min.pct filtering</span></span>
<span><span class="co">## 1194 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 283 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 6</span></span>
<span><span class="co">## 1130 features left after min.pct filtering</span></span>
<span><span class="co">## 1130 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 289 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 7</span></span>
<span><span class="co">## 1199 features left after min.pct filtering</span></span>
<span><span class="co">## 1199 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 189 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 8</span></span>
<span><span class="co">## 1154 features left after min.pct filtering</span></span>
<span><span class="co">## 1154 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 392 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 9</span></span>
<span><span class="co">## 1239 features left after min.pct filtering</span></span>
<span><span class="co">## 1239 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 363 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 10</span></span>
<span><span class="co">## 1473 features left after min.pct filtering</span></span>
<span><span class="co">## 1473 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 359 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 11</span></span>
<span><span class="co">## 1138 features left after min.pct filtering</span></span>
<span><span class="co">## 1138 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 280 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## -----------------------------------</span></span>
<span><span class="co">## testing cluster 12</span></span>
<span><span class="co">## 1208 features left after min.pct filtering</span></span>
<span><span class="co">## 1208 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 344 features left after log2fc.threshold filtering</span></span>
<span><span class="co">## -----------------------------------</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select the top 3 positive markers per cluster </span></span>
<span><span class="va">top3.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cluster.markers</span>, f <span class="op">=</span> <span class="va">cluster.markers</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">log2FC</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span>, n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">top3.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">top3.markers</span><span class="op">)</span></span>
<span><span class="va">top3.markers</span> <span class="op">&lt;-</span> <span class="va">top3.markers</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">top3.markers</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Heatmap of the top 3 positive markers per cluster</span></span>
<span><span class="fu"><a href="../reference/HeatmapFeatures.html">HeatmapFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, </span>
<span>                clustering.label <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>                features <span class="op">=</span> <span class="va">top3.markers</span><span class="op">$</span><span class="va">marker</span>, </span>
<span>                seed.color <span class="op">=</span> <span class="fl">65</span><span class="op">)</span></span></code></pre></div>
<p><img src="01_Integration_files/figure-html/cluster%20markers-1.png" width="480" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="dge">DGE<a class="anchor" aria-label="anchor" href="#dge"></a>
</h2>
<p><br></p>
<p><code>Coralysis</code> was able to separate the CD8 effector T cells
into two clusters: 6 and 11. From the differential gene expression (DGE)
analysis below, it is clear that cluster 11 is more cytotoxic and
similar to NK cells (expressing GZMH and GZMB) than cluster 6.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DGE analysis: cluster 6 vs 11</span></span>
<span><span class="va">dge.clt6vs11</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindClusterMarkers.html">FindClusterMarkers</a></span><span class="op">(</span><span class="va">pbmc_10Xassays</span>, </span>
<span>                                   clustering.label <span class="op">=</span> <span class="st">"cluster"</span>, </span>
<span>                                   clusters.1 <span class="op">=</span> <span class="st">"6"</span>, </span>
<span>                                   clusters.2 <span class="op">=</span> <span class="st">"11"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## testing cluster group.1</span></span>
<span><span class="co">## 997 features left after min.pct filtering</span></span>
<span><span class="co">## 997 features left after min.diff.pct filtering</span></span>
<span><span class="co">## 303 features left after log2fc.threshold filtering</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dge.clt6vs11</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">dge.clt6vs11</span><span class="op">$</span><span class="va">log2FC</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           p.value  adj.p.value    log2FC      pct.1     pct.2  diff.pct marker</span></span>
<span><span class="co">## NKG7 3.395687e-65 6.791373e-62 -4.087289 0.11403509 1.0000000 0.8859649   NKG7</span></span>
<span><span class="co">## CCL5 2.349768e-63 4.699536e-60 -3.838459 0.12573099 1.0000000 0.8742690   CCL5</span></span>
<span><span class="co">## GZMH 9.265926e-86 1.853185e-82 -3.170614 0.01461988 1.0000000 0.9853801   GZMH</span></span>
<span><span class="co">## CST7 1.986018e-69 3.972037e-66 -2.447930 0.04970760 0.9436620 0.8939544   CST7</span></span>
<span><span class="co">## GZMA 3.278546e-66 6.557091e-63 -2.417989 0.05263158 0.9154930 0.8628614   GZMA</span></span>
<span><span class="co">## LTB  2.631790e-33 5.263580e-30  2.325730 0.96491228 0.2253521 0.7395602    LTB</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top6.degs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dge.clt6vs11</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">dge.clt6vs11</span><span class="op">$</span><span class="va">log2FC</span><span class="op">)</span>, </span>
<span>                                     decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="st">"marker"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">exp.plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">top6.degs</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/PlotExpression.html">PlotExpression</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_10Xassays</span>, color.by <span class="op">=</span> <span class="va">x</span>,</span>
<span>                   scale.values <span class="op">=</span> <span class="cn">TRUE</span>, point.size <span class="op">=</span> <span class="fl">0.5</span>, point.stroke <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">exp.plots</span>, align <span class="op">=</span> <span class="st">"vh"</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="01_Integration_files/figure-html/dge-1.png" width="864" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="r-session">R session<a class="anchor" aria-label="anchor" href="#r-session"></a>
</h2>
<p><br></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># R session</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0</span></span>
<span><span class="co">##  [3] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">##  [7] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">##  [9] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">## [11] Coralysis_1.0.0             ggplot2_3.5.1              </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rlang_1.1.5              magrittr_2.0.3           flexclust_1.4-2         </span></span>
<span><span class="co">##   [4] compiler_4.4.2           systemfonts_1.2.1        vctrs_0.6.5             </span></span>
<span><span class="co">##   [7] reshape2_1.4.4           stringr_1.5.1            pkgconfig_2.0.3         </span></span>
<span><span class="co">##  [10] crayon_1.5.3             fastmap_1.2.0            XVector_0.46.0          </span></span>
<span><span class="co">##  [13] labeling_0.4.3           scuttle_1.16.0           rmarkdown_2.29          </span></span>
<span><span class="co">##  [16] ggbeeswarm_0.7.2         UCSC.utils_1.2.0         ragg_1.3.3              </span></span>
<span><span class="co">##  [19] xfun_0.50                modeltools_0.2-23        bluster_1.16.0          </span></span>
<span><span class="co">##  [22] zlibbioc_1.52.0          cachem_1.1.0             beachmat_2.22.0         </span></span>
<span><span class="co">##  [25] jsonlite_1.8.9           DelayedArray_0.32.0      BiocParallel_1.40.0     </span></span>
<span><span class="co">##  [28] irlba_2.3.5.1            parallel_4.4.2           aricode_1.0.3           </span></span>
<span><span class="co">##  [31] cluster_2.1.6            R6_2.6.0                 bslib_0.9.0             </span></span>
<span><span class="co">##  [34] stringi_1.8.4            RColorBrewer_1.1-3       limma_3.62.2            </span></span>
<span><span class="co">##  [37] jquerylib_0.1.4          Rcpp_1.0.14              iterators_1.0.14        </span></span>
<span><span class="co">##  [40] knitr_1.49               snow_0.4-4               Matrix_1.7-1            </span></span>
<span><span class="co">##  [43] igraph_2.1.4             tidyselect_1.2.1         abind_1.4-8             </span></span>
<span><span class="co">##  [46] yaml_2.3.10              codetools_0.2-20         doRNG_1.8.6.1           </span></span>
<span><span class="co">##  [49] lattice_0.22-6           tibble_3.2.1             plyr_1.8.9              </span></span>
<span><span class="co">##  [52] withr_3.0.2              ggrastr_1.0.2            Rtsne_0.17              </span></span>
<span><span class="co">##  [55] evaluate_1.0.3           desc_1.4.3               pillar_1.10.1           </span></span>
<span><span class="co">##  [58] rngtools_1.5.2           foreach_1.5.2            generics_0.1.3          </span></span>
<span><span class="co">##  [61] sparseMatrixStats_1.18.0 munsell_0.5.1            scales_1.3.0            </span></span>
<span><span class="co">##  [64] class_7.3-22             glue_1.8.0               metapod_1.14.0          </span></span>
<span><span class="co">##  [67] pheatmap_1.0.12          LiblineaR_2.10-24        tools_4.4.2             </span></span>
<span><span class="co">##  [70] BiocNeighbors_2.0.1      ScaledMatrix_1.14.0      SparseM_1.84-2          </span></span>
<span><span class="co">##  [73] RSpectra_0.16-2          locfit_1.5-9.11          RANN_2.6.2              </span></span>
<span><span class="co">##  [76] fs_1.6.5                 scran_1.34.0             Cairo_1.6-2             </span></span>
<span><span class="co">##  [79] cowplot_1.1.3            grid_4.4.2               edgeR_4.4.2             </span></span>
<span><span class="co">##  [82] colorspace_2.1-1         GenomeInfoDbData_1.2.13  beeswarm_0.4.0          </span></span>
<span><span class="co">##  [85] BiocSingular_1.22.0      vipor_0.4.7              cli_3.6.3               </span></span>
<span><span class="co">##  [88] rsvd_1.0.5               textshaping_1.0.0        viridisLite_0.4.2       </span></span>
<span><span class="co">##  [91] S4Arrays_1.6.0           dplyr_1.1.4              doSNOW_1.0.20           </span></span>
<span><span class="co">##  [94] gtable_0.3.6             sass_0.4.9               digest_0.6.37           </span></span>
<span><span class="co">##  [97] SparseArray_1.6.1        dqrng_0.4.1              farver_2.1.2            </span></span>
<span><span class="co">## [100] htmltools_0.5.8.1        pkgdown_2.1.1            lifecycle_1.0.4         </span></span>
<span><span class="co">## [103] httr_1.4.7               statmod_1.5.0</span></span></code></pre>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p><br></p>
<p>Amezquita R, Lun A, Becht E, Carey V, Carpp L, Geistlinger L, Marini
F, Rue-Albrecht K, Risso D, Soneson C, Waldron L, Pages H, Smith M,
Huber W, Morgan M, Gottardo R, Hicks S (2020). “Orchestrating
single-cell analysis with Bioconductor.” <em>Nature Methods</em>,
<em>17</em>, 137-145. <a href="https://www.nature.com/articles/s41592-019-0654-x" class="external-link">https://www.nature.com/articles/s41592-019-0654-x</a>.</p>
<p>Lun ATL, McCarthy DJ, Marioni JC (2016). “A step-by-step workflow for
low-level analysis of single-cell RNA-seq data with Bioconductor.”
<em>F1000Res.</em>, <em>5</em>, 2122. <a href="https://doi.org/10.12688/f1000research.9501.2" class="external-link">doi:10.12688/f1000research.9501.2</a>.</p>
<p>Sousa A, Smolander J, Junttila S, Elo L (2025). “Coralysis enables
sensitive identification of imbalanced cell types and states in
single-cell data via multi-level integration.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2025.02.07.637023" class="external-link">doi:10.1101/2025.02.07.637023</a></p>
<p>Wickham H (2016). “ggplot2: Elegant Graphics for Data Analysis.”
<em>Springer-Verlag New York</em>.</p>
<p><br></p>
<p><br></p>
<p><br></p>
<p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by António Sousa, Johannes Smolander, Sini Junttila, Laura L Elo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
