<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Reference-mapping • Coralysis</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reference-mapping">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Coralysis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/01_Integration.html">Integration</a></li>
    <li><a class="dropdown-item" href="../articles/02_RefMap.html">Reference-mapping</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/elolab/Coralysis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reference-mapping</h1>
            
            <h4 data-toc-skip class="date">Compiled: 11/02/2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/elolab/Coralysis/blob/main/vignettes/02_RefMap.Rmd" class="external-link"><code>vignettes/02_RefMap.Rmd</code></a></small>
      <div class="d-none name"><code>02_RefMap.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p><br></p>
<p>To illustrate the <code>Coralysis</code> reference-mapping method, we
will use in this vignette two 10X PBMCs (Peripheral Blood Mononuclear
Cells) 3’ assays: <strong>V1</strong> and <strong>V2</strong>. The assay
<strong>V2</strong> will be the <em>reference</em> and
<strong>V1</strong> the <em>query</em>, i.e., the dataset that will be
mapped against the reference. The datasets have been downloaded from 10X
website. The PBMC dataset <strong>V1</strong> corresponds to sample
<em>pbmc6k</em> and <strong>V2</strong> to <code>pbmc8k</code>:</p>
<ul>
<li><p><strong>V1</strong>: <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc6k" class="external-link">pbmc6k</a></p></li>
<li><p><strong>V2</strong>: <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k" class="external-link">pbmc8k</a></p></li>
</ul>
<p><br></p>
<p>Cells were annotated using the annotations provided by <a href="https://www.nature.com/articles/s41592-019-0619-0" class="external-link">Korsunsky et
al., 2019</a> (Source Data Figure 4 file). The overall data was
downsampled to 2K cells (1K per assay) and 2K highly variable genes
selected with <code>scran</code> R package. To facilitate the
reproduction of this vignette, the data is distributed through
<em>Zenodo</em> as a <code>SingleCellExperiment</code> object, the
object (class) required by most functions in <code>Coralysis</code> (see
<a href="https://bioconductor.org/books/3.13/OSCA.intro/the-singlecellexperiment-class.html" class="external-link">Chapter
4 The SingleCellExperiment class - OSCA manual</a>). The
<code>SCE</code> object provided comprises <code>counts</code> (raw
count data), <code>logcounts</code> (log-normalized data) and cell
<code>colData</code> (which includes batch and cell labels, designated
as <code>batch</code> and <code>cell_type</code>, respectively).</p>
<p>Run the code below to import the <code>R</code> packages and data
required to reproduce this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/elolab/Coralysis" class="external-link">"Coralysis"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Import data from Zenodo</span></span>
<span><span class="va">data.url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/14845751/files/pbmc_10Xassays.rds?download=1"</span></span>
<span><span class="va">pbmc_10Xassays</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="va">data.url</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Split the SCE object by assay </span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="va">pbmc_10Xassays</span><span class="op">[</span>,<span class="va">pbmc_10Xassays</span><span class="op">$</span><span class="va">batch</span><span class="op">==</span><span class="st">"V2"</span><span class="op">]</span> <span class="co"># let V2 assay batch be the reference data set</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="va">pbmc_10Xassays</span><span class="op">[</span>,<span class="va">pbmc_10Xassays</span><span class="op">$</span><span class="va">batch</span><span class="op">==</span><span class="st">"V1"</span><span class="op">]</span> <span class="co"># let V1 be the query (unknown annotations)</span></span></code></pre></div>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="train-reference">Train reference<a class="anchor" aria-label="anchor" href="#train-reference"></a>
</h2>
<p><br></p>
<p>The first step in performing reference mapping with
<code>Coralysis</code> is training a dataset that is representative of
the biological system under study. In this example, <code>ref</code> and
<code>query</code> correspond to the <code>SingleCellExperiment</code>
objects of the reference and query PBMC samples, <em>V2</em> and
<em>V1</em> 3’ assays, respectively. The reference <code>ref</code> is
trained through <code>RunParallelDivisiveICP</code> function without
providing a <code>batch.label</code>. In case the reference requires
integration, a <code>batch.label</code> should be provided. An higher
number of <code>threads</code> can be provided to speed up computing
time depending on the number of cores available. For this example, the
algorithm was run 10 times (<code>L = 10</code>), but generally, this
number should be higher (with the default being <code>L = 50</code>).
Next, run the function <code>RunPCA</code> to obtain the main result
required for cell type prediction later on. In addition to cell type
prediction, the query dataset(s) can be projected onto UMAP. To allow
this, the argument <code>return.model</code> should be set to
<code>TRUE</code> in both functions <code>RunPCA</code> and
<code>RunUMAP</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Train the reference </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunParallelDivisiveICP.html">RunParallelDivisiveICP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ref</span>, L <span class="op">=</span> <span class="fl">10</span>, threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># runs without 'batch.label' </span></span></code></pre></div>
<pre><code><span><span class="co">## WARNING: Setting 'divisive.method' to 'cluster' as 'batch.label=NULL'. </span></span>
<span><span class="co">## If 'batch.label=NULL', 'divisive.method' can be one of: 'cluster', 'random'.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Building training set...</span></span></code></pre>
<pre><code><span><span class="co">## Training set successfully built.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Computing cluster seed.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Initializing divisive ICP clustering...</span></span></code></pre>
<pre><code><span><span class="co">##   |                                                                              |                                                                      |   0%  |                                                                              |========                                                              |  11%  |                                                                              |================                                                      |  22%  |                                                                              |=======================                                               |  33%  |                                                                              |===============================                                       |  44%  |                                                                              |=======================================                               |  56%  |                                                                              |===============================================                       |  67%  |                                                                              |======================================================                |  78%  |                                                                              |==============================================================        |  89%  |                                                                              |======================================================================| 100%</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Divisive ICP clustering completed successfully.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Predicting cell cluster probabilities using ICP models...</span></span></code></pre>
<pre><code><span><span class="co">## Prediction of cell cluster probabilities completed successfully.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Multi-level integration completed successfully.</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute reference PCA</span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">ref</span>, return.model <span class="op">=</span> <span class="cn">TRUE</span>, pca.method <span class="op">=</span> <span class="st">"stats"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Divisive ICP: selecting ICP tables multiple of 4</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute reference UMAP</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ref</span>, return.model <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Below is the UMAP plot for the reference sample with cell type
annotations. In this example, we are using the annotations provided in
the object. Ideally, the sample should be annotated after training by
performing clustering and manual cluster annotation. Only the resulting
manual cell type annotations should be used for prediction. For
simplicity, we will use the annotations provided in the object.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Vizualize reference</span></span>
<span><span class="va">ref.celltype.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ref</span>, </span>
<span>                                color.by <span class="op">=</span> <span class="st">"cell_type"</span>, </span>
<span>                                dimred <span class="op">=</span> <span class="st">"UMAP"</span>, </span>
<span>                                point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                                legend.nrow <span class="op">=</span> <span class="fl">6</span>, </span>
<span>                                seed.color <span class="op">=</span> <span class="fl">7</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"reference (ground-truth)"</span><span class="op">)</span></span>
<span><span class="va">ref.celltype.plot</span></span></code></pre></div>
<p><img src="02_RefMap_files/figure-html/viz%20reference-1.png" width="288" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="map-query">Map query<a class="anchor" aria-label="anchor" href="#map-query"></a>
</h2>
<p><br></p>
<p>Perform reference-mapping with <code>Coralysis</code> by running the
function <code>ReferenceMapping</code>. This requires to provide the
trained reference (<code>ref</code>) with cell type annotations intended
for the prediction (<code>ref.label = "cell_type"</code>) and the query
dataset (<code>query</code>). The label in the reference aimed to be
used for prediction needs to be available on <code>colData(ref)</code>.
In this case, we are providing the cell type labels from the column
<code>cell_type</code> available in the reference <code>ref</code>.
Since we want to project the query onto the reference UMAP, set
<code>project.umap</code> as <code>TRUE</code>. The argument
<code>dimred.name.prefix</code> just sets the name given as prefix of
the low dimensional embeddings stored in
<code>reducedDimNames(map)</code>. The <code>SingleCellExperiment</code>
object <code>map</code> will contain the same information as
<code>query</code>, with the predictions and embeddings mapped onto the
reference. The predictions consist in <code>coral_labels</code> and
<code>coral_probability</code> stored in <code>colData(map)</code>. The
<code>coral_labels</code> correspond to the cell type predictions
obtained against the reference. The <code>coral_probability</code>
represents the proportion of <em>K</em> neighbors from the winning class
(<code>k.nn</code> equal 10 by default); the higher the value, the
better.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Reference-mapping</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1024</span><span class="op">)</span></span>
<span><span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReferenceMapping.html">ReferenceMapping</a></span><span class="op">(</span>ref <span class="op">=</span> <span class="va">ref</span>, query <span class="op">=</span> <span class="va">query</span>, ref.label <span class="op">=</span> <span class="st">"cell_type"</span>, </span>
<span>                        project.umap <span class="op">=</span> <span class="cn">TRUE</span>, dimred.name.prefix <span class="op">=</span> <span class="st">"ref"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="prediction-accuracy">Prediction accuracy<a class="anchor" aria-label="anchor" href="#prediction-accuracy"></a>
</h2>
<p><br></p>
<p>The accuracy of <code>Coralysis</code> reference-mapping method is
presented below together with a confusion matrix between the predicted
(rows) versus ground-truth cell type labels (columns).</p>
<p><br></p>
<p><br></p>
<div class="section level3">
<h3 id="confusion-matrix">Confusion matrix<a class="anchor" aria-label="anchor" href="#confusion-matrix"></a>
</h3>
<p><br></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Confusion matrix</span></span>
<span><span class="va">preds_x_truth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">map</span><span class="op">$</span><span class="va">coral_labels</span>, <span class="va">map</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">preds_x_truth</span><span class="op">)</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">preds_x_truth</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Accuracy</span></span>
<span><span class="va">acc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">preds_x_truth</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">preds_x_truth</span><span class="op">)</span></span>
<span><span class="co">#print(paste0("Prediction accuracy: ", acc*100, "%"))</span></span>
<span></span>
<span><span class="co"># Print confusion matrix</span></span>
<span><span class="va">preds_x_truth</span></span></code></pre></div>
<pre><code><span><span class="co">##                 </span></span>
<span><span class="co">##                  aDC B mem B naive CD4 mem CD4 naive CD8 eff CD8 T HSC</span></span>
<span><span class="co">##   aDC             14     0       0       0         0       0     0   0</span></span>
<span><span class="co">##   B mem            0    40      15       0         0       0     0   3</span></span>
<span><span class="co">##   B naive          0     4      69       0         0       0     0   0</span></span>
<span><span class="co">##   CD4 mem          0     0       0     108        39       3     6   0</span></span>
<span><span class="co">##   CD4 naive        0     0       0      14       153       0    12   2</span></span>
<span><span class="co">##   CD8 eff          0     1       0       2         0      93     2   0</span></span>
<span><span class="co">##   CD8 T            0     0       1       3         4      11    62   0</span></span>
<span><span class="co">##   HSC              0     0       0       0         0       0     0   0</span></span>
<span><span class="co">##   Megakaryocyte    0     0       0       0         0       0     0   0</span></span>
<span><span class="co">##   Monocyte         1     0       0       0         0       0     0   0</span></span>
<span><span class="co">##   CD16+ monocyte   0     0       0       0         0       0     0   0</span></span>
<span><span class="co">##   NK               0     0       0       0         0       1     0   0</span></span>
<span><span class="co">##   pDC              0     0       0       0         0       0     0   0</span></span>
<span><span class="co">##   Treg             0     0       0       1         0       0     0   0</span></span>
<span><span class="co">##                 </span></span>
<span><span class="co">##                  Megakaryocyte Monocyte CD16+ monocyte  NK pDC Treg</span></span>
<span><span class="co">##   aDC                        0        0              0   0   0    0</span></span>
<span><span class="co">##   B mem                      0        0              0   0   0    0</span></span>
<span><span class="co">##   B naive                    0        0              0   0   1    0</span></span>
<span><span class="co">##   CD4 mem                    0        0              0   0   0    8</span></span>
<span><span class="co">##   CD4 naive                  0        0              0   0   0    7</span></span>
<span><span class="co">##   CD8 eff                    1        0              0   1   0    0</span></span>
<span><span class="co">##   CD8 T                      1        0              0   0   0    0</span></span>
<span><span class="co">##   HSC                        0        0              0   0   0    0</span></span>
<span><span class="co">##   Megakaryocyte              0        0              0   0   0    0</span></span>
<span><span class="co">##   Monocyte                   3      173              4   1   0    0</span></span>
<span><span class="co">##   CD16+ monocyte             1        5             72   0   0    0</span></span>
<span><span class="co">##   NK                         0        0              0  55   0    0</span></span>
<span><span class="co">##   pDC                        0        0              0   0   3    0</span></span>
<span><span class="co">##   Treg                       0        0              0   0   0    0</span></span></code></pre>
<p><br></p>
<p>The accuracy of <code>Coralysis</code> reference-mapping method was
84.2%.</p>
<p><br></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="dimred">DimRed<a class="anchor" aria-label="anchor" href="#dimred"></a>
</h3>
<p><br></p>
<p>Visualize below the query cells projected onto reference UMAP and how
well the predictions match the query ground-truth. The
<code>coral_probability</code> is a prediction confidence score.
Predictions with low scores (&lt;0.5) should be carefully inspected.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot query and reference UMAP side-by-side </span></span>
<span><span class="co">#with ground-truth &amp; predicted cell labels</span></span>
<span><span class="va">use.color</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"aDC"</span> <span class="op">=</span> <span class="st">"#E6F5C9"</span>, </span>
<span>               <span class="st">"B mem"</span> <span class="op">=</span> <span class="st">"#CCEBC5"</span>, </span>
<span>               <span class="st">"B naive"</span> <span class="op">=</span> <span class="st">"#FB8072"</span>, </span>
<span>               <span class="st">"CD4 mem"</span> <span class="op">=</span> <span class="st">"#A6761D"</span>, </span>
<span>               <span class="st">"CD4 naive"</span> <span class="op">=</span> <span class="st">"#666666"</span>, </span>
<span>               <span class="st">"CD8 eff"</span> <span class="op">=</span> <span class="st">"#80B1D3"</span>,</span>
<span>               <span class="st">"CD8 T"</span> <span class="op">=</span> <span class="st">"#CBD5E8"</span>, </span>
<span>               <span class="st">"HSC"</span> <span class="op">=</span> <span class="st">"#E31A1C"</span>, </span>
<span>               <span class="st">"Megakaryocyte"</span> <span class="op">=</span> <span class="st">"#377EB8"</span>, </span>
<span>               <span class="st">"Monocyte"</span> <span class="op">=</span> <span class="st">"#FCCDE5"</span>, </span>
<span>               <span class="st">"CD16+ monocyte"</span> <span class="op">=</span> <span class="st">"#A6D854"</span>, </span>
<span>               <span class="st">"NK"</span> <span class="op">=</span> <span class="st">"#6A3D9A"</span>,</span>
<span>               <span class="st">"pDC"</span> <span class="op">=</span> <span class="st">"#E7298A"</span>, </span>
<span>               <span class="st">"Treg"</span> <span class="op">=</span> <span class="st">"#FFFF33"</span><span class="op">)</span></span>
<span><span class="va">query.ground_truth.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">map</span>, </span>
<span>                                      color.by <span class="op">=</span> <span class="st">"cell_type"</span>, </span>
<span>                                      dimred <span class="op">=</span> <span class="st">"refUMAP"</span>, </span>
<span>                                      point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                                      legend.nrow <span class="op">=</span> <span class="fl">6</span>, </span>
<span>                                      seed.color <span class="op">=</span> <span class="fl">7</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"query (ground-truth)"</span><span class="op">)</span></span>
<span><span class="va">query.predicted.plot</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">map</span>, </span>
<span>                                  color.by <span class="op">=</span> <span class="st">"coral_labels"</span>, </span>
<span>                                  dimred <span class="op">=</span> <span class="st">"refUMAP"</span>, point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                                  legend.nrow <span class="op">=</span> <span class="fl">6</span>, </span>
<span>                                  use.color <span class="op">=</span> <span class="va">use.color</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"query (predicted)"</span><span class="op">)</span></span>
<span><span class="va">query.confidence.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotExpression.html">PlotExpression</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">map</span>, </span>
<span>                                        color.by <span class="op">=</span> <span class="st">"coral_probability"</span>, </span>
<span>                                        dimred <span class="op">=</span> <span class="st">"refUMAP"</span>, </span>
<span>                                        point.size <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                                        color.scale <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"query (confidence)"</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">ref.celltype.plot</span>, <span class="va">query.ground_truth.plot</span>, </span>
<span>                   <span class="va">query.predicted.plot</span>, <span class="va">query.confidence.plot</span>, </span>
<span>                   ncol <span class="op">=</span> <span class="fl">2</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span></span></code></pre></div>
<p><img src="02_RefMap_files/figure-html/prediction%20viz-1.png" width="696" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="r-session">R session<a class="anchor" aria-label="anchor" href="#r-session"></a>
</h3>
<p><br></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># R session</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0</span></span>
<span><span class="co">##  [3] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">##  [7] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">##  [9] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">## [11] Coralysis_1.0.0             ggplot2_3.5.1              </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rlang_1.1.5              magrittr_2.0.3           flexclust_1.4-2         </span></span>
<span><span class="co">##   [4] compiler_4.4.2           png_0.1-8                systemfonts_1.2.1       </span></span>
<span><span class="co">##   [7] vctrs_0.6.5              reshape2_1.4.4           stringr_1.5.1           </span></span>
<span><span class="co">##  [10] pkgconfig_2.0.3          crayon_1.5.3             fastmap_1.2.0           </span></span>
<span><span class="co">##  [13] XVector_0.46.0           labeling_0.4.3           scuttle_1.16.0          </span></span>
<span><span class="co">##  [16] rmarkdown_2.29           ggbeeswarm_0.7.2         UCSC.utils_1.2.0        </span></span>
<span><span class="co">##  [19] ragg_1.3.3               xfun_0.50                modeltools_0.2-23       </span></span>
<span><span class="co">##  [22] bluster_1.16.0           zlibbioc_1.52.0          cachem_1.1.0            </span></span>
<span><span class="co">##  [25] beachmat_2.22.0          jsonlite_1.8.9           DelayedArray_0.32.0     </span></span>
<span><span class="co">##  [28] BiocParallel_1.40.0      irlba_2.3.5.1            parallel_4.4.2          </span></span>
<span><span class="co">##  [31] aricode_1.0.3            cluster_2.1.6            R6_2.5.1                </span></span>
<span><span class="co">##  [34] bslib_0.9.0              stringi_1.8.4            RColorBrewer_1.1-3      </span></span>
<span><span class="co">##  [37] reticulate_1.40.0        limma_3.62.2             jquerylib_0.1.4         </span></span>
<span><span class="co">##  [40] Rcpp_1.0.14              iterators_1.0.14         knitr_1.49              </span></span>
<span><span class="co">##  [43] snow_0.4-4               Matrix_1.7-1             igraph_2.1.4            </span></span>
<span><span class="co">##  [46] tidyselect_1.2.1         abind_1.4-8              yaml_2.3.10             </span></span>
<span><span class="co">##  [49] codetools_0.2-20         doRNG_1.8.6.1            lattice_0.22-6          </span></span>
<span><span class="co">##  [52] tibble_3.2.1             plyr_1.8.9               withr_3.0.2             </span></span>
<span><span class="co">##  [55] askpass_1.2.1            ggrastr_1.0.2            evaluate_1.0.3          </span></span>
<span><span class="co">##  [58] desc_1.4.3               pillar_1.10.1            rngtools_1.5.2          </span></span>
<span><span class="co">##  [61] foreach_1.5.2            generics_0.1.3           sparseMatrixStats_1.18.0</span></span>
<span><span class="co">##  [64] munsell_0.5.1            scales_1.3.0             class_7.3-22            </span></span>
<span><span class="co">##  [67] glue_1.8.0               metapod_1.14.0           pheatmap_1.0.12         </span></span>
<span><span class="co">##  [70] LiblineaR_2.10-24        tools_4.4.2              BiocNeighbors_2.0.1     </span></span>
<span><span class="co">##  [73] ScaledMatrix_1.14.0      SparseM_1.84-2           RSpectra_0.16-2         </span></span>
<span><span class="co">##  [76] locfit_1.5-9.11          RANN_2.6.2               fs_1.6.5                </span></span>
<span><span class="co">##  [79] scran_1.34.0             Cairo_1.6-2              cowplot_1.1.3           </span></span>
<span><span class="co">##  [82] grid_4.4.2               umap_0.2.10.0            edgeR_4.4.2             </span></span>
<span><span class="co">##  [85] colorspace_2.1-1         GenomeInfoDbData_1.2.13  beeswarm_0.4.0          </span></span>
<span><span class="co">##  [88] BiocSingular_1.22.0      vipor_0.4.7              cli_3.6.3               </span></span>
<span><span class="co">##  [91] rsvd_1.0.5               textshaping_1.0.0        viridisLite_0.4.2       </span></span>
<span><span class="co">##  [94] S4Arrays_1.6.0           dplyr_1.1.4              doSNOW_1.0.20           </span></span>
<span><span class="co">##  [97] gtable_0.3.6             sass_0.4.9               digest_0.6.37           </span></span>
<span><span class="co">## [100] SparseArray_1.6.1        dqrng_0.4.1              farver_2.1.2            </span></span>
<span><span class="co">## [103] htmltools_0.5.8.1        pkgdown_2.1.1            lifecycle_1.0.4         </span></span>
<span><span class="co">## [106] httr_1.4.7               statmod_1.5.0            openssl_2.3.2</span></span></code></pre>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<p><br></p>
<p>Amezquita R, Lun A, Becht E, Carey V, Carpp L, Geistlinger L, Marini
F, Rue-Albrecht K, Risso D, Soneson C, Waldron L, Pages H, Smith M,
Huber W, Morgan M, Gottardo R, Hicks S (2020). “Orchestrating
single-cell analysis with Bioconductor.” <em>Nature Methods</em>,
<em>17</em>, 137-145. <a href="https://www.nature.com/articles/s41592-019-0654-x" class="external-link">https://www.nature.com/articles/s41592-019-0654-x</a>.</p>
<p>Sousa A, Smolander J, Junttila S, Elo L (2025). “Coralysis enables
sensitive identification of imbalanced cell types and states in
single-cell data via multi-level integration.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2025.02.07.637023" class="external-link">doi:10.1101/2025.02.07.637023</a></p>
<p>Wickham H (2016). “ggplot2: Elegant Graphics for Data Analysis.”
<em>Springer-Verlag New York</em>.</p>
<p><br></p>
<p><br></p>
<p><br></p>
<p><br></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by António Sousa, Johannes Smolander, Sini Junttila, Laura L Elo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
