<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Cell states • Coralysis</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Cell states">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Coralysis</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/Coralysis.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/CellState.html">Cell states</a></li>
    <li><a class="dropdown-item" href="../articles/Integration.html">Integration</a></li>
    <li><a class="dropdown-item" href="../articles/RefMap.html">Reference-mapping</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/elolab/Coralysis/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Cell states</h1>
            
            <h4 data-toc-skip class="date">Compiled: 14/02/2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/elolab/Coralysis/blob/main/vignettes/CellState.Rmd" class="external-link"><code>vignettes/CellState.Rmd</code></a></small>
      <div class="d-none name"><code>CellState.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p><br></p>
<p>The single-cell RNA sequencing dataset of peripheral blood
mononuclear cells (PBMCs) from healthy versus COVID-19 infected donors
published by <a href="https://www.nature.com/articles/s41591-020-0944-y" class="external-link">Wilk et al.,
(2020)</a> will be used to demonstrated how cell cluster probability can
be used to identify perturbed cell states and associated differential
gene expression programs. In addition to the original cell type
annotations, the object provided contains the cell type annotations
given by <a href="https://www.frontiersin.org/journals/genetics/articles/10.3389/fgene.2022.929887/full" class="external-link">Välikangas
et al., 2022</a>.</p>
<p>The original dataset has been downsampled to 21K cells (1.5K cells
per sample; from 44,721 cells) to increase the speed and reduce the
computational resources required to run this vignette. To facilitate the
reproduction of this vignette, the data is distributed through
<em>Zenodo</em> as a <code>SingleCellExperiment</code> object, the
object (class) required by most functions in <code>Coralysis</code> (see
<a href="https://bioconductor.org/books/3.13/OSCA.intro/the-singlecellexperiment-class.html" class="external-link">Chapter
4 The SingleCellExperiment class - OSCA manual</a>)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install packages </span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="st">"scater"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pkg_install.html" class="external-link">pkg_install</a></span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="st">"ComplexHeatmap"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pkg_install.html" class="external-link">pkg_install</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/scater/" class="external-link">"scater"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/elolab/Coralysis" class="external-link">"Coralysis"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Import data </span></span>
<span><span class="va">data.url</span> <span class="op">&lt;-</span> <span class="st">"https://zenodo.org/records/14845751/files/covid_Wilk_et_al.rds?download=1"</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="va">data.url</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Downsample every sample to 1.5K cells: 21K cells in total</span></span>
<span><span class="va">cells.by.donor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, f <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">Donor_full</span><span class="op">)</span></span>
<span><span class="va">ncells</span> <span class="op">&lt;-</span> <span class="fl">1.5e3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">down.cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">cells.by.donor</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, size <span class="op">=</span> <span class="va">ncells</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">down.cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">down.cells</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">down.cells</span><span class="op">]</span></span></code></pre></div>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h2>
<p><br></p>
<p><code>Coralysis</code> requires log-normalised data as input. The
dataset above has been previously normalised with the basic
log-normalisation method in <code>Seurat</code>. Below is a simple
custom function that can be used to perform <code>Seurat</code> (see <a href="https://satijalab.org/seurat/reference/normalizedata" class="external-link">NormalizeData</a>).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Normalize the data</span></span>
<span><span class="co"># log1p normalization</span></span>
<span><span class="va">SeuratNormalisation</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># 'SeuratNormalisation()' function applies the basic Seurat normalization to</span></span>
<span>  <span class="co">#a SingleCellExperiment object with a 'counts' assay. Normalized data</span></span>
<span>  <span class="co">#is saved in the 'logcounts' assay.</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">x</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">*</span><span class="fl">10000</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span> <span class="co"># log1p normalization w/ 10K scaling factor</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span>, <span class="st">"sparseMatrix"</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">SeuratNormalisation</span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p>In alternative, <code>scran</code> normalization can be performed,
which is particularly beneficial if rare cell types exist (see the
following <a href="https://bioconductor.org/packages/devel/bioc/vignettes/scran/inst/doc/scran.html" class="external-link">vignette</a>).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## scran normalisation</span></span>
<span><span class="va">ScranNormalisation</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">norm_clusters</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span>    <span class="va">object</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">object</span>, clusters <span class="op">=</span> <span class="va">norm_clusters</span><span class="op">)</span></span>
<span>    <span class="va">object</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Split object by batch: 'Donor_full'</span></span>
<span><span class="va">batch.levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">Donor_full</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">batch.levels</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">batch.levels</span></span>
<span><span class="va">sce.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">batch.levels</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">sce</span><span class="op">$</span><span class="va">Donor_full</span> <span class="op">==</span> <span class="va">x</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Apply normalisation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sce.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sce.list</span>, <span class="va">ScranNormalisation</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Join </span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="va">sce.list</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="hvg-selection">HVG selection<a class="anchor" aria-label="anchor" href="#hvg-selection"></a>
</h2>
<p><br></p>
<p>Highly variable genes (HVG) can be selected using the R package
<code>scran</code>. The variable <code>Donor_full</code> from
<code>colData</code> is used as batch label. The
<code>SingleCellExperiment</code> object allows alternative experiments
to the main experiment. This is important to keep a backup of all genes
in the same <code>SingleCellExperiment</code> object before selecting
HVGs (see <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html" class="external-link">SingleCellExperiment
vignette</a>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Feature selection with 'scran' package</span></span>
<span><span class="va">nhvg</span> <span class="op">&lt;-</span> <span class="fl">2000</span></span>
<span><span class="va">batch.label</span> <span class="op">&lt;-</span> <span class="st">"Donor_full"</span></span>
<span><span class="va">sce</span><span class="op">[[</span><span class="va">batch.label</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[[</span><span class="va">batch.label</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">m.hvg</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span>, block<span class="op">=</span><span class="va">sce</span><span class="op">[[</span><span class="va">batch.label</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">hvg.ordered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">m.hvg</span><span class="op">[[</span><span class="st">"bio"</span><span class="op">]</span><span class="op">]</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">top.hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="va">hvg.ordered</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">nhvg</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="st">"highly_variable"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">top.hvg</span>, <span class="va">m.hvg</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset object </span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="va">top.hvg</span>,<span class="op">]</span></span></code></pre></div>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="dimred-pre-integration">DimRed: pre-integration<a class="anchor" aria-label="anchor" href="#dimred-pre-integration"></a>
</h2>
<p><br></p>
<p>A PCA and UMAP representation of dataset before performing
integration can be obtained with <code>Coralysis</code>. Several
parameters can be provided to the functions in order to adjust the
analyses to the dataset (see <code>?Coralysis::RunPCA()</code> and
<code><a href="../reference/RunUMAP.html">?Coralysis::RunUMAP</a></code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Dimensional reduction - unintegrated</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunPCA.html">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, assay.name <span class="op">=</span> <span class="st">"logcounts"</span>, dimred.name <span class="op">=</span> <span class="st">"unintPCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, dimred.type <span class="op">=</span> <span class="st">"unintPCA"</span>, </span>
<span>               umap.method <span class="op">=</span> <span class="st">"uwot"</span>, dimred.name <span class="op">=</span> <span class="st">"unintUMAP"</span>, </span>
<span>               n_neighbors <span class="op">=</span> <span class="fl">15</span>, min_dist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting </span></span>
<span><span class="va">batch.label</span> <span class="op">&lt;-</span> <span class="st">"Donor_full"</span></span>
<span><span class="va">original.cell.label</span> <span class="op">&lt;-</span> <span class="st">"cell_type"</span></span>
<span><span class="va">cell.label</span> <span class="op">&lt;-</span> <span class="st">"predicted_celltype_l2"</span></span>
<span><span class="va">vars2plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">batch.label</span>, <span class="va">original.cell.label</span>, <span class="va">cell.label</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vars2plot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">vars2plot</span></span>
<span><span class="va">unint.plts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">vars2plot</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, color.by <span class="op">=</span> <span class="va">x</span>, dimred <span class="op">=</span> <span class="st">"unintUMAP"</span>, point.size <span class="op">=</span> <span class="fl">0.2</span>, </span>
<span>               legend.nrow <span class="op">=</span> <span class="fl">10</span>, point.stroke <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="co"># plot each variable</span></span>
<span><span class="va">all.unint.plts</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">unint.plts</span>, ncol <span class="op">=</span> <span class="fl">3</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span> <span class="co"># join plots together</span></span>
<span><span class="va">all.unint.plts</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/plot%20unintegrated%20dimred-1.png" width="1008" style="display: block; margin: auto;"></p>
<p>It is clear the donor effect, particularly for some cell types, such
as CD14 monocytes and CD8 memory T cells.</p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="multi-level-integration">Multi-level integration<a class="anchor" aria-label="anchor" href="#multi-level-integration"></a>
</h2>
<p><br></p>
<p>Multi-level integration can be performed with <code>Coralysis</code>
by running the function <code><a href="../reference/RunParallelDivisiveICP.html">RunParallelDivisiveICP()</a></code>. Only the
<code>batch.label</code> parameter is required. The
<code>batch.label</code> corresponds to a variable in
<code>colData</code>. In this case <code>Donor_full</code>. The
<code><a href="../reference/RunParallelDivisiveICP.html">RunParallelDivisiveICP()</a></code> function can be run in parallel by
providing the number of <code>threads</code> to reduce the run time (it
takes around 25 min. with ICP batch size 1K cells and 4 threads).
Consult the full documentation of this function with
<code><a href="../reference/RunParallelDivisiveICP.html">?RunParallelDivisiveICP</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform multi-level integration</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunParallelDivisiveICP.html">RunParallelDivisiveICP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, batch.label <span class="op">=</span> <span class="st">"Donor_full"</span>, </span>
<span>                              icp.batch.size <span class="op">=</span> <span class="fl">1000</span>, threads <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><code>Coralysis</code> returns a list of cell cluster probabilities
saved at <code>metadata(sce)$coralysis$joint.probability</code> of
length equal to the number of icp runs, i.e., <code>L</code> (by default
<code>L=50</code>), times the number of icp rounds, i.e.,
log2(<code>k</code>) (by default <code>k=16</code>, thus 4 rounds of
divisive icp).</p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="icp-clusters">ICP clusters<a class="anchor" aria-label="anchor" href="#icp-clusters"></a>
</h2>
<p><br></p>
<p>The cell cluster probability for a given icp run through its divisive
rounds can be plotted with the function <code><a href="../reference/PlotClusterTree.html">PlotClusterTree()</a></code>.
A <code>colData</code> variable can be provided to see its composition
per cluster across divisive rounds for the respective icp run. See
examples below for icp run number 16.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot cluster tree for icp run 16</span></span>
<span><span class="co"># Probability</span></span>
<span><span class="va">prob.cluster.tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotClusterTree.html">PlotClusterTree</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, icp.run <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># batch label distribution</span></span>
<span><span class="va">batch.cluster.tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotClusterTree.html">PlotClusterTree</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, icp.run <span class="op">=</span> <span class="fl">16</span>, color.by <span class="op">=</span> <span class="st">"Donor_full"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># original cell type label distribution</span></span>
<span><span class="va">orig.cell.cluster.tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotClusterTree.html">PlotClusterTree</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, icp.run <span class="op">=</span> <span class="fl">16</span>, color.by <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Azimuth/predicted cell type label distribution</span></span>
<span><span class="va">cell.cluster.tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotClusterTree.html">PlotClusterTree</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, icp.run <span class="op">=</span> <span class="fl">16</span>, color.by <span class="op">=</span> <span class="st">"predicted_celltype_l2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Join all plots with 'cowplot'</span></span>
<span><span class="va">all.cluster.tree</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">prob.cluster.tree</span>, <span class="va">batch.cluster.tree</span>, </span>
<span>                                                          ncol <span class="op">=</span> <span class="fl">2</span>, rel_widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span>, align <span class="op">=</span> <span class="st">"v"</span><span class="op">)</span>, </span>
<span>                                       <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">orig.cell.cluster.tree</span>, <span class="va">cell.cluster.tree</span>, </span>
<span>                                                          ncol <span class="op">=</span> <span class="fl">2</span>, rel_widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span>,</span>
<span>                                       ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">all.cluster.tree</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/multi-level%20probability-1.png" width="1344" style="display: block; margin: auto;"></p>
<p>Some cell types formed unique clusters, but quite few remained mixed.
Multi-level integration resolution might get better by running
<code><a href="../reference/RunParallelDivisiveICP.html">RunParallelDivisiveICP()</a></code> with 32 cluster, i.e.,
<code>k = 32</code>, instead 16.</p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="dimred-after-integration">DimRed: after integration<a class="anchor" aria-label="anchor" href="#dimred-after-integration"></a>
</h2>
<p><br></p>
<p><code>Coralysis</code> returns a list of cell cluster probabilities
saved at <code>metadata(sce)$coralysis$joint.probability</code> of
length equal to the number of icp runs, i.e., <code>L</code> (by default
<code>L=50</code>), times the number of icp rounds, i.e.,
log2(<code>k</code>) (by default <code>k=16</code>, thus 4 rounds of
divisive icp). The cell cluster probability can be concatenated to
compute a PCA in order to obtain an integrated embedded. By default only
the cell cluster probability corresponding to the last icp round is
used. The <code>Coralysis</code> integrated embedding can be used
downstream for non-linear dimensional reduction, <em>t</em>-SNE or UMAP,
and clustering.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Dimensional reduction - unintegrated</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunPCA.html">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, assay.name <span class="op">=</span> <span class="st">"joint.probability"</span>, dimred.name <span class="op">=</span> <span class="st">"intPCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, dimred.type <span class="op">=</span> <span class="st">"intPCA"</span>, </span>
<span>               umap.method <span class="op">=</span> <span class="st">"uwot"</span>, dimred.name <span class="op">=</span> <span class="st">"intUMAP"</span>, </span>
<span>               dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, n_neighbors <span class="op">=</span> <span class="fl">15</span>, min_dist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting </span></span>
<span><span class="va">int.plts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">vars2plot</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, color.by <span class="op">=</span> <span class="va">x</span>, dimred <span class="op">=</span> <span class="st">"intUMAP"</span>, point.size <span class="op">=</span> <span class="fl">0.2</span>, legend.nrow <span class="op">=</span> <span class="fl">10</span>, point.stroke <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="co"># plot each variable</span></span>
<span><span class="va">all.int.plts</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">int.plts</span>, ncol <span class="op">=</span> <span class="fl">3</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span> <span class="co"># join plots together</span></span>
<span><span class="va">all.int.plts</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/integrated%20plots-1.png" width="1008" style="display: block; margin: auto;"></p>
<p><code>Coralysis</code> efficiently integrated the donor PBMC samples,
identified small populations (e.g., neutrophils, Treg, MAIT, dnT, NK
CD56bright), and differentiation trajectories (e.g., B naive to
memory).</p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="graph-based-clustering">Graph-based clustering<a class="anchor" aria-label="anchor" href="#graph-based-clustering"></a>
</h2>
<p><br></p>
<p>Graph-based clustering can be obtained by running the
<code>scran</code> function <code>clusterCells()</code> using the
<code>Coralysis</code> integrated embedding. In alternative, the joint
cluster probabilities can be used for clustering. The advantages of
using the integrated embedding instead joint probabilities are:
computational efficiency; and, noise robustness. Using the joint
probabilities instead PCA embedding takes considerably more time.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Graph-based clustering with scran</span></span>
<span><span class="co">## Coralysis integrated PCA embedding </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"cltPCA"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"intPCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1024</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">emb_clusters</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"cltPCA"</span>, </span>
<span>                                        BLUSPARAM <span class="op">=</span> <span class="fu">bluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">15</span>, cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Coralysis joint cluster probabilities</span></span>
<span><span class="co"># retrieve ICP cell cluster probability tables for every icp run, but only for the last divisive round</span></span>
<span><span class="va">probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetCellClusterProbability.html">GetCellClusterProbability</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, icp.round <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> </span>
<span><span class="co"># dim(probs) # 21K cells x 800 clusters (16 clusters x 50 icp runs = 800 clusters) </span></span>
<span><span class="va">probs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">probs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">probs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">prob.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"probability"</span> <span class="op">=</span> <span class="va">probs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1024</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">prob_clusters</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">prob.sce</span>, </span>
<span>                                         assay.type <span class="op">=</span> <span class="st">"probability"</span>,</span>
<span>                                         BLUSPARAM <span class="op">=</span> <span class="fu">bluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">15</span>, cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting</span></span>
<span><span class="va">vars2plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">cell.label</span>, <span class="st">"emb_clusters"</span>, <span class="st">"prob_clusters"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vars2plot2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">vars2plot2</span></span>
<span><span class="va">clts.plts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">vars2plot2</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/PlotDimRed.html">PlotDimRed</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"intUMAP"</span>, color.by <span class="op">=</span> <span class="va">x</span>, point.size <span class="op">=</span> <span class="fl">0.2</span>, legend.nrow <span class="op">=</span> <span class="fl">10</span>, point.stroke <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="co"># plot each variable</span></span>
<span><span class="va">all.clts.plts</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">clts.plts</span>, ncol <span class="op">=</span> <span class="fl">3</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span> <span class="co"># join plots together</span></span>
<span><span class="va">all.clts.plts</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/clustering%20plot-1.png" width="1008" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="cell-state-identification">Cell state identification<a class="anchor" aria-label="anchor" href="#cell-state-identification"></a>
</h2>
<p><br></p>
<p>The cell cluster probability aggregated by mean or median across the
icp runs can be obtained with the function
<code><a href="../reference/SummariseCellClusterProbability.html">SummariseCellClusterProbability()</a></code> and plotted to infer
transient and steady cell states.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summarise cell cluster probability</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SummariseCellClusterProbability.html">SummariseCellClusterProbability</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, icp.round <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="co"># save result in 'colData'</span></span>
<span><span class="co"># colData(sce) # check the colData</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot cell cluster probabilities - mean</span></span>
<span><span class="co"># possible options: "mean_probs", "median_probs", "scaled_median_probs" </span></span>
<span><span class="fu"><a href="../reference/PlotExpression.html">PlotExpression</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, color.by <span class="op">=</span> <span class="st">"scaled_mean_probs"</span>, dimred <span class="op">=</span> <span class="st">"intUMAP"</span>, color.scale <span class="op">=</span> <span class="st">"viridis"</span>, </span>
<span>               point.size <span class="op">=</span> <span class="fl">0.2</span>, point.stroke <span class="op">=</span> <span class="fl">0.1</span>, legend.title <span class="op">=</span> <span class="st">"Mean prob.\n(min-max)"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="CellState_files/figure-html/plot%20probability-1.png" width="480" style="display: block; margin: auto;"></p>
<p>For instance, B cell states (naive, intermediate, memory) are not
well supported by <code>Coralysis</code> cell cluster probability,
probably because it would require to run <code>Coralysis</code> at
higher resolution, i.e., <code>k=32</code> instead
<code>k=16</code>.</p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="gene-coefficients">Gene coefficients<a class="anchor" aria-label="anchor" href="#gene-coefficients"></a>
</h2>
<p><br></p>
<p>Gene coefficients can be obtained for a given cell label through
majority voting with the function <code><a href="../reference/MajorityVotingFeatures.html">MajorityVotingFeatures()</a></code>.
The majority voting is performed by searching for the most
representative cluster for a given cell label across all possible
clusters (i.e., across all icp runs and rounds). The most representative
cluster corresponds to the cluster with the highest majority voting
score. This corresponds to the geometric mean between the proportion of
cells from the given label intersected with a cluster and the proportion
of cells from that same cluster that intersected with the cells from the
given label. The higher the score the better. A cluster that scores 1
indicates that all its cells correspond to the assigned label, and vice
versa; i.e., all the cells with the assigned label belong to this
cluster. For example, <code><a href="../reference/MajorityVotingFeatures.html">MajorityVotingFeatures()</a></code> by providing
the <code>colData</code> variable <code>"emb_clusters"</code>(i.e.,
<code>label="emb_clusters"</code>).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get label gene coefficients by majority voting</span></span>
<span><span class="va">clts.gene.coeff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MajorityVotingFeatures.html">MajorityVotingFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, label <span class="op">=</span> <span class="st">"emb_clusters"</span><span class="op">)</span></span></code></pre></div>
<p>The function <code><a href="../reference/MajorityVotingFeatures.html">MajorityVotingFeatures()</a></code> returns a list
with two elements:</p>
<ul>
<li><p><code>feature_coeff</code>: a list of data frames comprising the
gene coefficients and respective weights.</p></li>
<li><p><code>summary</code>: a data frame summarizing which ICP cluster
best represents each cell label, along with the respective majority
voting score.</p></li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Print cluster gene coefficients summary</span></span>
<span><span class="va">clts.gene.coeff</span><span class="op">$</span><span class="va">summary</span></span></code></pre></div>
<pre><code><span><span class="co">##    label icp_run icp_round cluster     score</span></span>
<span><span class="co">## 1      1      22         4       8 0.8157863</span></span>
<span><span class="co">## 2      2      24         4      16 0.9421878</span></span>
<span><span class="co">## 3      3      21         4      11 0.5959206</span></span>
<span><span class="co">## 4      4      21         4      14 0.6799433</span></span>
<span><span class="co">## 5      5      29         4       7 0.6721712</span></span>
<span><span class="co">## 6      6      48         4      11 0.7086803</span></span>
<span><span class="co">## 7      7      36         4      14 0.4917519</span></span>
<span><span class="co">## 8      8      16         4       2 0.8996775</span></span>
<span><span class="co">## 9      9      22         4       7 0.5562097</span></span>
<span><span class="co">## 10    10      47         4      16 0.6487200</span></span>
<span><span class="co">## 11    11      19         4       4 0.9326988</span></span>
<span><span class="co">## 12    12      24         4       9 0.5858352</span></span>
<span><span class="co">## 13    13      21         4       2 0.7723912</span></span>
<span><span class="co">## 14    14      41         4       6 0.8999670</span></span>
<span><span class="co">## 15    15      37         4       3 0.8040199</span></span>
<span><span class="co">## 16    16      29         4      10 0.6894886</span></span>
<span><span class="co">## 17    17      48         4      12 0.9532055</span></span>
<span><span class="co">## 18    18      46         4       5 0.7898323</span></span>
<span><span class="co">## 19    19      18         4      10 0.6683786</span></span>
<span><span class="co">## 20    20      20         4       7 0.9176726</span></span>
<span><span class="co">## 21    21      31         4       1 0.8014923</span></span></code></pre>
<p><br></p>
<p><br></p>
<div class="section level3">
<h3 id="unique-marker-prss23">Unique marker: <em>PRSS23</em><a class="anchor" aria-label="anchor" href="#unique-marker-prss23"></a>
</h3>
<p><br></p>
<p>One of the strengths of <code>Coralyis</code> consist of finding
unique cluster markers as it was the case for cluster 2.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Print top 30 positive gene coefficients for cluster 2: NK + CD8 TEM </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">clts.gene.coeff</span><span class="op">$</span><span class="va">feature_coeff</span><span class="op">$</span><span class="va">`2`</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">clts.gene.coeff</span><span class="op">$</span><span class="va">feature_coeff</span><span class="op">$</span><span class="va">`2`</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span>, <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     feature   coeff_clt16</span></span>
<span><span class="co">## 43   PRSS23  4.3504965756</span></span>
<span><span class="co">## 18     GZMB  0.1637480659</span></span>
<span><span class="co">## 19     GZMH  0.1308493859</span></span>
<span><span class="co">## 79   ZBTB44  0.1109175583</span></span>
<span><span class="co">## 40    PATL2  0.0983686833</span></span>
<span><span class="co">## 72    TBX21  0.0954275398</span></span>
<span><span class="co">## 7     CD247  0.0753691598</span></span>
<span><span class="co">## 15     GNLY  0.0730897748</span></span>
<span><span class="co">## 61    RUNX3  0.0686593222</span></span>
<span><span class="co">## 39    PARP8  0.0654860566</span></span>
<span><span class="co">## 42     PRF1  0.0504557585</span></span>
<span><span class="co">## 8      CDK6  0.0427174844</span></span>
<span><span class="co">## 27    KLRD1  0.0402690820</span></span>
<span><span class="co">## 17     GZMA  0.0375472215</span></span>
<span><span class="co">## 46      PXN  0.0360506239</span></span>
<span><span class="co">## 14     GNG2  0.0223021975</span></span>
<span><span class="co">## 25   IPCEF1  0.0156802538</span></span>
<span><span class="co">## 45   PTGER4  0.0087374973</span></span>
<span><span class="co">## 6      CCL5  0.0082154040</span></span>
<span><span class="co">## 47    RBM27  0.0031892948</span></span>
<span><span class="co">## 80     ZEB2  0.0018210680</span></span>
<span><span class="co">## 69  ST3GAL1 -0.0005403014</span></span>
<span><span class="co">## 12   GIMAP4 -0.0008745088</span></span>
<span><span class="co">## 68  SMARCA2 -0.0013411496</span></span>
<span><span class="co">## 21 HSP90AB1 -0.0020678336</span></span>
<span><span class="co">## 75   TMSB4X -0.0027057304</span></span>
<span><span class="co">## 51    RPL34 -0.0032341377</span></span>
<span><span class="co">## 57     RPS3 -0.0033493263</span></span>
<span><span class="co">## 4     BIRC6 -0.0040249087</span></span>
<span><span class="co">## 2      ACTB -0.0047227878</span></span></code></pre>
<p><br></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the expression of PRSS23</span></span>
<span><span class="fu"><a href="../reference/PlotExpression.html">PlotExpression</a></span><span class="op">(</span><span class="va">sce</span>, color.by <span class="op">=</span> <span class="st">"PRSS23"</span>, dimred <span class="op">=</span> <span class="st">"intUMAP"</span>, point.size <span class="op">=</span> <span class="fl">0.5</span>, point.stroke <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/plot%20PRSS23-1.png" width="480" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="cluster-markers">Cluster markers<a class="anchor" aria-label="anchor" href="#cluster-markers"></a>
</h2>
<p><br></p>
<div class="section level3">
<h3 id="cd14-mono-1-vs-5">CD14 Mono: 1 vs 5<a class="anchor" aria-label="anchor" href="#cd14-mono-1-vs-5"></a>
</h3>
<p><br></p>
<p><code>Coralysis</code> identified three main clusters of CD14
monocytes: 1 (2,028 cells), 5 (1,651), and 9 (935) (considering the
Louvain clustering obtained with the integrated embedding). The clusters
1 was compared against 5 below.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cluster gene coefficients: 1 versus 8</span></span>
<span><span class="va">clt1</span> <span class="op">&lt;-</span> <span class="va">clts.gene.coeff</span><span class="op">$</span><span class="va">feature_coeff</span><span class="op">$</span><span class="va">`1`</span></span>
<span><span class="va">clt1</span> <span class="op">&lt;-</span> <span class="va">clt1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">clt1</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">clt5</span> <span class="op">&lt;-</span> <span class="va">clts.gene.coeff</span><span class="op">$</span><span class="va">feature_coeff</span><span class="op">$</span><span class="va">`5`</span></span>
<span><span class="va">clt5</span> <span class="op">&lt;-</span> <span class="va">clt5</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">clt5</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Merge cluster coefficients</span></span>
<span><span class="va">clt1vs5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">clt1</span>, <span class="va">clt5</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">clt1vs5</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">clt1vs5</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">clt1vs5.parsed</span> <span class="op">&lt;-</span> <span class="va">clt1vs5</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="st">"coeff_variation"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">coeff_clt7</span> <span class="op">-</span> <span class="va">coeff_clt8</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">coeff_variation</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">coeff_clt7</span><span class="op">!=</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">coeff_clt8</span><span class="op">!=</span><span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">(</span><span class="va">coeff_clt7</span> <span class="op">*</span> <span class="va">coeff_clt8</span><span class="op">)</span><span class="op">&lt;</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">top5.clt1</span> <span class="op">&lt;-</span> <span class="va">clt1vs5.parsed</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">coeff_clt8</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">top5.clt5</span> <span class="op">&lt;-</span> <span class="va">clt1vs5.parsed</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">coeff_clt7</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Top 5 positive coefficients in each cluster</span></span>
<span><span class="va">top5.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">top5.clt1</span>, <span class="va">top5.clt5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">top5.genes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">top5.genes</span></span>
<span></span>
<span><span class="co"># Plot</span></span>
<span><span class="va">top5.gexp.mono.clts.plts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">top5.genes</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/PlotExpression.html">PlotExpression</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, color.by <span class="op">=</span> <span class="va">x</span>, dimred <span class="op">=</span> <span class="st">"intUMAP"</span>, </span>
<span>                   scale.values <span class="op">=</span> <span class="cn">TRUE</span>, point.size <span class="op">=</span> <span class="fl">0.25</span>, point.stroke <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">all.top5.gexp.mono.clts.plts</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">top5.gexp.mono.clts.plts</span>, ncol <span class="op">=</span> <span class="fl">5</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span></span>
<span><span class="va">all.top5.gexp.mono.clts.plts</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/cluster%20gene%20coefficients:%201%20vs%205-1.png" width="1536" style="display: block; margin: auto;"></p>
<p><br></p>
<p>The expression level of CYP1B1 in cluster 1 is slightly higher in
COVID-19 infected donor cells than healthy donors, particularly, for
ventilated versus healthy donors.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">sce</span><span class="op">$</span><span class="va">emb_clusters</span><span class="op">==</span><span class="st">"1"</span><span class="op">]</span>, features <span class="op">=</span> <span class="st">"CYP1B1"</span>, </span>
<span>                       color_by <span class="op">=</span> <span class="st">"Ventilated"</span>, x <span class="op">=</span> <span class="st">"Ventilated"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/violin%20plot:%20CYP1B1-1.png" width="288" style="display: block; margin: auto;"></p>
<p><br></p>
<p>The previous plots suggest that clusters 1 and 5 are phenotypically
different. In addition, low expression of HLA-DR (HLA-DRA, HLA-DRB1)
genes has been associated with severe inflammatory conditions like
COVID-19 (see plots below) as well as high expression of S100A9. These
results suggest that cluster 1 corresponds to strongly activated or
possibly dysregulated monocytes, which are more common in severe
COVID-19 than in cluster 5. Indeed, 40% the cells in cluster 1 are from
COVID-19 infected donors that required ventilation, whereas this
percentage was around to 25% for cluster 5.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes2plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"HLA-DRA"</span>, <span class="st">"HLA-DRB1"</span><span class="op">)</span></span>
<span><span class="va">gexp.hla.plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">genes2plot</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/PlotExpression.html">PlotExpression</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, color.by <span class="op">=</span> <span class="va">x</span>, dimred <span class="op">=</span> <span class="st">"intUMAP"</span>,  </span>
<span>                   scale.values <span class="op">=</span> <span class="cn">TRUE</span>, point.size <span class="op">=</span> <span class="fl">0.25</span>, point.stroke <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">gexp.hla.plots</span>, ncol <span class="op">=</span> <span class="fl">2</span>, align <span class="op">=</span> <span class="st">"vh"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/plot%20HLA-DR%20genes-1.png" width="768" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="perturbed-cell-states">Perturbed cell states<a class="anchor" aria-label="anchor" href="#perturbed-cell-states"></a>
</h2>
<p><br></p>
<div class="section level3">
<h3 id="cd16-monocytes">CD16 Monocytes<a class="anchor" aria-label="anchor" href="#cd16-monocytes"></a>
</h3>
<p><br></p>
<p>The cell cluster probability can be used to search for altered cell
states, e.g., associated with COVID-19. The distribution of cell cluster
probability per cluster (integrated embedding) per
<code>ventilated</code> group is shown below after running the function
<code><a href="../reference/CellClusterProbabilityDistribution.html">CellClusterProbabilityDistribution()</a></code>. Cluster 14, roughly
corresponding to CD16 monocytes, shows a difference between healthy and
COVID associated cells (non-ventilated and ventilated donor cells).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Disease associated cell state </span></span>
<span><span class="co"># Search for diseased affected cell states</span></span>
<span><span class="va">prob.dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CellClusterProbabilityDistribution.html">CellClusterProbabilityDistribution</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, </span>
<span>                                                label <span class="op">=</span> <span class="st">"emb_clusters"</span>, </span>
<span>                                                group <span class="op">=</span> <span class="st">"Ventilated"</span>, </span>
<span>                                                probability <span class="op">=</span> <span class="st">"scaled_mean_probs"</span><span class="op">)</span></span>
<span><span class="va">prob.dist</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/disease-associated%20cell%20states-1.png" width="1920" style="display: block; margin: auto;"></p>
<p><br></p>
<p>Cell cluster probability bins per cluster <code>label</code> can be
obtained by running <code><a href="../reference/BinCellClusterProbability.html">BinCellClusterProbability()</a></code>, which
returns a <code>SingleCellExperiment</code> object with
<code>logcounts</code> average per <code>label</code> per cell
probability bin. The number of probability bins can be provided to the
parameter <code>bins</code>. The cell frequency per cell bin per group
of interest can be obtained with the function
<code><a href="../reference/TabulateCellBinsByGroup.html">TabulateCellBinsByGroup()</a></code>. The heatmap below shows the cell
frequency across ventilated groups per bins for cluster 14
(corresponding to CD16 monocytes). The cell frequency represented below
corresponds to the cell proportions per group (healthy, non-ventilated
and ventilated) scaled by bins, i.e., columns. Cell states with higher
probabilities tend to be more “healthier” than states with lower
probabilities.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cell states SCE object </span></span>
<span><span class="va">cellstate.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BinCellClusterProbability.html">BinCellClusterProbability</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">sce</span>, label <span class="op">=</span> <span class="st">"emb_clusters"</span>, icp.round <span class="op">=</span> <span class="fl">4</span>, bins <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Tabulate cell bins by group </span></span>
<span><span class="va">cellbins.tables</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TabulateCellBinsByGroup.html">TabulateCellBinsByGroup</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellstate.sce</span>, group <span class="op">=</span> <span class="st">"Ventilated"</span>, relative <span class="op">=</span> <span class="cn">TRUE</span>, margin <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Heatmap of cell bins distribution by group for cluster 20 - CD16 Monocytes</span></span>
<span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="st">"14"</span></span>
<span><span class="va">col.fun</span> <span class="op">&lt;-</span> <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">0</span>, <span class="fl">1.5</span><span class="op">)</span>, <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">inferno</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">heat.plot</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">cellbins.tables</span><span class="op">[[</span><span class="va">cluster</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                                     name <span class="op">=</span> <span class="st">"Col. Z-score"</span>, </span>
<span>                                     cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                     cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                     row_names_side <span class="op">=</span> <span class="st">"left"</span>, </span>
<span>                                     col <span class="op">=</span> <span class="va">col.fun</span>, </span>
<span>                                     column_title <span class="op">=</span> <span class="st">"Distribution of cells per cell cluster probability bin"</span>, </span>
<span>                                     heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>legend_direction <span class="op">=</span> <span class="st">"horizontal"</span>, </span>
<span>                                                                 title_position <span class="op">=</span> <span class="st">"topcenter"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">heat.plot</span>, heatmap_legend_side <span class="op">=</span> <span class="st">"bottom"</span>, annotation_legend_side <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/tabulate%20cell%20bins%20by%20group-1.png" width="480" style="display: block; margin: auto;"></p>
<p>The correlation between cell cluster probability bins from cluster 18
and averaged gene expression can be obtained with the function
<code><a href="../reference/CellBinsFeatureCorrelation.html">CellBinsFeatureCorrelation()</a></code>. Among the several positively
and negatively correlated genes, the expression of C1QB seems to be
enriched in diseased cells, with a negative Pearson coefficient close to
0.7. This gene was selected as an example due to its expression being
relatively limited to cluster 14.</p>
<p><br></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Pearson correlated features with cluster 24</span></span>
<span><span class="va">cor.features.clt14</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CellBinsFeatureCorrelation.html">CellBinsFeatureCorrelation</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cellstate.sce</span>, labels <span class="op">=</span> <span class="va">cluster</span><span class="op">)</span></span>
<span><span class="va">cor.features.clt14</span> <span class="op">&lt;-</span> <span class="va">cor.features.clt14</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">`14`</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">`14`</span><span class="op">)</span> <span class="co"># C1QB</span></span>
<span><span class="va">exp.C1QB.umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PlotExpression.html">PlotExpression</a></span><span class="op">(</span><span class="va">sce</span>, color.by <span class="op">=</span> <span class="st">"C1QB"</span>, dimred <span class="op">=</span> <span class="st">"intUMAP"</span>, </span>
<span>                                scale.values <span class="op">=</span> <span class="cn">TRUE</span>, point.size <span class="op">=</span> <span class="fl">0.75</span>, point.stroke <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="va">exp.C1QB.violin</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[</span>,<span class="va">sce</span><span class="op">$</span><span class="va">emb_clusters</span> <span class="op">==</span> <span class="va">cluster</span><span class="op">]</span>, features <span class="op">=</span> <span class="st">"C1QB"</span>,</span>
<span>                                          x <span class="op">=</span> <span class="st">"Ventilated"</span>, color_by <span class="op">=</span> <span class="st">"Ventilated"</span><span class="op">)</span></span>
<span><span class="va">exp.C1QB.umap</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/differential%20expression%20programs-1.png" width="384" style="display: block; margin: auto;"></p>
<p><br></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp.C1QB.violin</span></span></code></pre></div>
<p><img src="CellState_files/figure-html/print%20violin-1.png" width="700" style="display: block; margin: auto;"></p>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="r-session">R session<a class="anchor" aria-label="anchor" href="#r-session"></a>
</h2>
<p><br></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># R session</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] Coralysis_1.0.0             scater_1.34.0              </span></span>
<span><span class="co">##  [3] ggplot2_3.5.1               scuttle_1.16.0             </span></span>
<span><span class="co">##  [5] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0</span></span>
<span><span class="co">##  [7] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">##  [9] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">## [11] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">## [13] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">## [15] dplyr_1.1.4                </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3       shape_1.4.6.1            jsonlite_1.8.9          </span></span>
<span><span class="co">##   [4] magrittr_2.0.3           modeltools_0.2-23        ggbeeswarm_0.7.2        </span></span>
<span><span class="co">##   [7] farver_2.1.2             rmarkdown_2.29           GlobalOptions_0.1.2     </span></span>
<span><span class="co">##  [10] fs_1.6.5                 zlibbioc_1.52.0          ragg_1.3.3              </span></span>
<span><span class="co">##  [13] vctrs_0.6.5              Cairo_1.6-2              htmltools_0.5.8.1       </span></span>
<span><span class="co">##  [16] S4Arrays_1.6.0           BiocNeighbors_2.0.1      SparseArray_1.6.1       </span></span>
<span><span class="co">##  [19] sass_0.4.9               bslib_0.9.0              desc_1.4.3              </span></span>
<span><span class="co">##  [22] plyr_1.8.9               cachem_1.1.0             igraph_2.1.4            </span></span>
<span><span class="co">##  [25] lifecycle_1.0.4          iterators_1.0.14         pkgconfig_2.0.3         </span></span>
<span><span class="co">##  [28] rsvd_1.0.5               Matrix_1.7-1             R6_2.6.0                </span></span>
<span><span class="co">##  [31] fastmap_1.2.0            clue_0.3-66              GenomeInfoDbData_1.2.13 </span></span>
<span><span class="co">##  [34] aricode_1.0.3            digest_0.6.37            colorspace_2.1-1        </span></span>
<span><span class="co">##  [37] dqrng_0.4.1              RSpectra_0.16-2          irlba_2.3.5.1           </span></span>
<span><span class="co">##  [40] textshaping_1.0.0        beachmat_2.22.0          labeling_0.4.3          </span></span>
<span><span class="co">##  [43] httr_1.4.7               polyclip_1.10-7          abind_1.4-8             </span></span>
<span><span class="co">##  [46] compiler_4.4.2           rngtools_1.5.2           doParallel_1.0.17       </span></span>
<span><span class="co">##  [49] withr_3.0.2              BiocParallel_1.40.0      viridis_0.6.5           </span></span>
<span><span class="co">##  [52] ggforce_0.4.2            LiblineaR_2.10-24        MASS_7.3-61             </span></span>
<span><span class="co">##  [55] DelayedArray_0.32.0      rjson_0.2.23             bluster_1.16.0          </span></span>
<span><span class="co">##  [58] tools_4.4.2              vipor_0.4.7              beeswarm_0.4.0          </span></span>
<span><span class="co">##  [61] scatterpie_0.2.4         flexclust_1.4-2          glue_1.8.0              </span></span>
<span><span class="co">##  [64] grid_4.4.2               cluster_2.1.6            reshape2_1.4.4          </span></span>
<span><span class="co">##  [67] generics_0.1.3           snow_0.4-4               gtable_0.3.6            </span></span>
<span><span class="co">##  [70] class_7.3-22             tidyr_1.3.1              BiocSingular_1.22.0     </span></span>
<span><span class="co">##  [73] ScaledMatrix_1.14.0      metapod_1.14.0           XVector_0.46.0          </span></span>
<span><span class="co">##  [76] ggrepel_0.9.6            RANN_2.6.2               foreach_1.5.2           </span></span>
<span><span class="co">##  [79] pillar_1.10.1            stringr_1.5.1            yulab.utils_0.2.0       </span></span>
<span><span class="co">##  [82] limma_3.62.2             circlize_0.4.16          tweenr_2.0.3            </span></span>
<span><span class="co">##  [85] lattice_0.22-6           SparseM_1.84-2           tidyselect_1.2.1        </span></span>
<span><span class="co">##  [88] ComplexHeatmap_2.22.0    locfit_1.5-9.11          knitr_1.49              </span></span>
<span><span class="co">##  [91] gridExtra_2.3            edgeR_4.4.2              xfun_0.50               </span></span>
<span><span class="co">##  [94] statmod_1.5.0            pheatmap_1.0.12          stringi_1.8.4           </span></span>
<span><span class="co">##  [97] UCSC.utils_1.2.0         ggfun_0.1.8              yaml_2.3.10             </span></span>
<span><span class="co">## [100] evaluate_1.0.3           codetools_0.2-20         tibble_3.2.1            </span></span>
<span><span class="co">## [103] cli_3.6.4                systemfonts_1.2.1        munsell_0.5.1           </span></span>
<span><span class="co">## [106] jquerylib_0.1.4          Rcpp_1.0.14              doSNOW_1.0.20           </span></span>
<span><span class="co">## [109] png_0.1-8                ggrastr_1.0.2            parallel_4.4.2          </span></span>
<span><span class="co">## [112] pkgdown_2.1.1            doRNG_1.8.6.1            scran_1.34.0            </span></span>
<span><span class="co">## [115] sparseMatrixStats_1.18.0 viridisLite_0.4.2        scales_1.3.0            </span></span>
<span><span class="co">## [118] purrr_1.0.4              crayon_1.5.3             GetoptLong_1.0.5        </span></span>
<span><span class="co">## [121] rlang_1.1.5              cowplot_1.1.3</span></span></code></pre>
<p><br></p>
<p><br></p>
<hr>
<p><br></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p><br></p>
<p>Amezquita R, Lun A, Becht E, Carey V, Carpp L, Geistlinger L, Marini
F, Rue-Albrecht K, Risso D, Soneson C, Waldron L, Pages H, Smith M,
Huber W, Morgan M, Gottardo R, Hicks S (2020). “Orchestrating
single-cell analysis with Bioconductor.” <em>Nature Methods</em>,
<em>17</em>, 137-145. <a href="https://www.nature.com/articles/s41592-019-0654-x" class="external-link">https://www.nature.com/articles/s41592-019-0654-x</a>.</p>
<p>Gu, Z. (2022) “Complex heatmap visualization.” <em>iMeta</em>.
1(3):e43. <a href="https://doi.org/10.1002/imt2.43" class="external-link">doi:
10.1002/imt2.43</a>.</p>
<p>Lun ATL, McCarthy DJ, Marioni JC (2016). “A step-by-step workflow for
low-level analysis of single-cell RNA-seq data with Bioconductor.”
<em>F1000Res.</em>, <em>5</em>, 2122. <a href="https://doi.org/10.12688/f1000research.9501.2" class="external-link">doi:10.12688/f1000research.9501.2</a>.</p>
<p>McCarthy DJ, Campbell KR, Lun ATL, Willis QF (2017). “Scater:
pre-processing, quality control, normalisation and visualisation of
single-cell RNA-seq data in R.” <em>Bioinformatics</em>, <em>33</em>,
1179-1186. <a href="https://doi.org/10.1093/bioinformatics/btw777" class="external-link">doi:10.1093/bioinformatics/btw777</a>.</p>
<p>Sousa A, Smolander J, Junttila S, Elo L (2025). “Coralysis enables
sensitive identification of imbalanced cell types and states in
single-cell data via multi-level integration.” <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2025.02.07.637023" class="external-link">doi:10.1101/2025.02.07.637023</a></p>
<p>Wickham H (2016). “ggplot2: Elegant Graphics for Data Analysis.”
<em>Springer-Verlag New York</em>.</p>
<p><br></p>
<p><br></p>
<p><br></p>
<p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by António Sousa, Johannes Smolander, Sini Junttila, Laura L Elo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
